<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Visualizing data with LLMs • scplotter</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Visualizing data with LLMs">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scplotter</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Visualizing_data_with_LLMs.html">Visualizing data with LLMs</a></li>
    <li><a class="dropdown-item" href="../articles/Working_with_anndata_h5ad_files.html">Working with anndata (.h5ad) files</a></li>
    <li><a class="dropdown-item" href="../articles/Knowing_your_spatial_data_and_visualization.html">Knowing your spatial data and visualization</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-spatial-examples" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Spatial Examples</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-spatial-examples">
<li><a class="dropdown-item" href="../articles/Seurat_10x_Visium.html">Visualizing 10x Visium data prepared with Seurat</a></li>
    <li><a class="dropdown-item" href="../articles/Seurat_10x_VisiumHD.html">Visualizing 10x VisiumHD data prepared with Seurat</a></li>
    <li><a class="dropdown-item" href="../articles/Seurat_SlideSeq.html">Visualizing SlideSeq data prepared with Seurat</a></li>
    <li><a class="dropdown-item" href="../articles/Seurat_Xenium.html">Visualizing Xenium data prepared with Seurat</a></li>
    <li><a class="dropdown-item" href="../articles/Seurat_Nanostring_CosMx.html">Visualizing Nanostring CosMx data prepared with Seurat</a></li>
    <li><a class="dropdown-item" href="../articles/Seurat_Akoya_CODEX.html">Visualizing Akoya_CODEX data prepared with Seurat</a></li>
    <li><a class="dropdown-item" href="../articles/Giotto_Visium.html">Visualizing Visium data prepared with Giotto</a></li>
    <li><a class="dropdown-item" href="../articles/Giotto_VisiumHD.html">Visualizing VisiumHD data prepared with Giotto</a></li>
    <li><a class="dropdown-item" href="../articles/Giotto_Xenium.html">Visualizing Xenium data prepared with Giotto</a></li>
    <li><a class="dropdown-item" href="../articles/Giotto_SlideSeq.html">Visualizing SlideSeq data prepared with Giotto</a></li>
    <li><a class="dropdown-item" href="../articles/Giotto_Spatial_CITE-Seq.html">Visualizing Spatial CITE-Seq data prepared with Giotto</a></li>
    <li><a class="dropdown-item" href="../articles/Giotto_Nanostring_CosMx.html">Visualizing Nanostirng CosMx data prepared with Giotto</a></li>
    <li><a class="dropdown-item" href="../articles/Giotto_CODEX.html">Visualizing CODEX data prepared with Giotto</a></li>
    <li><a class="dropdown-item" href="../articles/Giotto_vizgen.html">Visualizing vizgen data prepared with Giotto</a></li>
    <li><a class="dropdown-item" href="../articles/Giotto_seqFISH.html">Visualizing seqFISH data prepared with Giotto</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pwwang/scplotter/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Visualizing data with LLMs</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/pwwang/scplotter/blob/master/vignettes/Visualizing_data_with_LLMs.Rmd" class="external-link"><code>vignettes/Visualizing_data_with_LLMs.Rmd</code></a></small>
      <div class="d-none name"><code>Visualizing_data_with_LLMs.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pwwang/scplotter" class="external-link">scplotter</a></span><span class="op">)</span></span>
<span><span class="va">api_key_set</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">"OPENAI_API_KEY"</span><span class="op">)</span>, <span class="st">""</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette demonstrates how to use the <code>scplotter</code>
package to visualize data with AI. The package provides a variety of
functions for visualizing single-cell sequencing data, including
scRNA-seq and scTCR-seq/scBCR-seq data.</p>
</div>
<div class="section level2">
<h2 id="setup-llm-provider">Setup LLM provider<a class="anchor" aria-label="anchor" href="#setup-llm-provider"></a>
</h2>
<p><code>scplotter</code> uses <code>tidyprompt</code> to provide a
unified interface for different LLM providers. You can set up your
preferred LLM provider using one of the <a href="https://tjarkvandemerwe.github.io/tidyprompt/reference/index.html#llm-providers-chat-history" class="external-link">wrappers</a>
provided by <code>tidyprompt</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set up LLM provider</span></span>
<span><span class="va">provider</span> <span class="op">&lt;-</span> <span class="fu">tidyprompt</span><span class="fu">::</span><span class="fu"><a href="https://kennispunttwente.github.io/tidyprompt/reference/llm_provider_openai.html" class="external-link">llm_provider_openai</a></span><span class="op">(</span></span>
<span>    parameters <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>model <span class="op">=</span> <span class="st">"gpt-5-nano"</span>, stream <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"tidyprompt.stream"</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    verbose <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"tidyprompt.verbose"</span>, <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    url <span class="op">=</span> <span class="st">"https://api.openai.com/v1/chat/completions"</span>,</span>
<span>    api_key <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">"OPENAI_API_KEY"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">chat</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/SCPlotterChat.html">SCPlotterChat</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>provider <span class="op">=</span> <span class="va">provider</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="setup-the-data-for-visualization">Setup the data for visualization<a class="anchor" aria-label="anchor" href="#setup-the-data-for-visualization"></a>
</h2>
<p>By default, <code>chat</code> will detects the data used for
visualization from the <code>.GlobalEnv</code> and data exported from
the <code>Seurat</code>, <code>SeuratObject</code>, and
<code>scRepertoire</code> packages.</p>
<p>You can also ask to list the available data:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chat</span><span class="op">$</span><span class="fu">ask</span><span class="op">(</span><span class="st">"List the available data that can be used for visualization."</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Tool identified:  ListData </span></span>
<span><span class="co">#&gt; Available data objects:</span></span>
<span><span class="co">#&gt; -  scplotter::cellphonedb_res :  A toy example of CellPhoneDB output from LIANA </span></span>
<span><span class="co">#&gt; -  scplotter::ifnb_sub :  A subsetted version of 'ifnb' datasets </span></span>
<span><span class="co">#&gt; -  scplotter::pancreas_sub :  A subsetted version of mouse 'pancreas' datasets </span></span>
<span><span class="co">#&gt; -  Seurat::cc.genes :  Cell cycle genes </span></span>
<span><span class="co">#&gt; -  Seurat::cc.genes.updated.2019 :  Cell cycle genes: 2019 update </span></span>
<span><span class="co">#&gt; -  SeuratObject::pbmc_small :  A small example version of the PBMC dataset </span></span>
<span><span class="co">#&gt; -  scRepertoire::contig_list :  A list of 8 single-cell T cell receptor sequences runs. </span></span>
<span><span class="co">#&gt; -  scRepertoire::mini_contig_list :  Processed subset of 'contig_list' </span></span>
<span><span class="co">#&gt; -  scRepertoire::scRep_example :  A Seurat object of 500 single T cells,</span></span>
<span><span class="co"># or you can do it explicitly</span></span>
<span><span class="co"># chat$list_data()</span></span></code></pre></div>
<p>To set up the data manually, you can use the <code>set_data()</code>
method.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chat</span><span class="op">$</span><span class="fu">set_data</span><span class="op">(</span><span class="fu">scplotter</span><span class="fu">::</span><span class="va"><a href="../reference/cellphonedb_res.html">cellphonedb_res</a></span><span class="op">)</span></span>
<span><span class="co"># To let the LLM to detect the data from the prompt again:</span></span>
<span><span class="va">chat</span><span class="op">$</span><span class="fu">set_data</span><span class="op">(</span><span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<p>To use your own data, you can either set the data manually or use the
<code>set_data()</code> method or you can load the data in the global
environment and mention it in your prompt.</p>
</div>
<div class="section level2">
<h2 id="list-the-available-tools">List the available tools<a class="anchor" aria-label="anchor" href="#list-the-available-tools"></a>
</h2>
<p>You can list the available functions by using the
<code>list_tools()</code> method.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chat</span><span class="op">$</span><span class="fu">list_tools</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Available tools:</span></span>
<span><span class="co">#&gt; -  gt :  Helper functions to select clones based on various criteria     </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; -  CellDimPlot :  Cell Dimension Reduction Plot     </span></span>
<span><span class="co">#&gt;    This function creates a dimension reduction plot for a Seurat object</span></span>
<span><span class="co">#&gt;    a Giotto object, a path to an .h5ad file or an opened    H5File    by    hdf5r    package.</span></span>
<span><span class="co">#&gt;    It allows for various customizations such as grouping by metadata,</span></span>
<span><span class="co">#&gt;    adding edges between cell neighbors, highlighting specific cells, and more.</span></span>
<span><span class="co">#&gt;    This function is a wrapper around    plotthis::DimPlot()   , which provides a</span></span>
<span><span class="co">#&gt;    flexible way to visualize cell clusters in reduced dimensions. This function</span></span>
<span><span class="co">#&gt;    extracts the necessary data from the Seurat or Giotto object and passes it to</span></span>
<span><span class="co">#&gt;    plotthis::DimPlot()   .</span></span>
<span><span class="co">#&gt; -  ClonalVolumePlot :  ClonalVolumePlot     </span></span>
<span><span class="co">#&gt;    ClonalVolumePlot</span></span>
<span><span class="co">#&gt; -  ClonalRarefactionPlot :  ClonalRarefactionPlot     </span></span>
<span><span class="co">#&gt;    Plot the rarefaction curves</span></span>
<span><span class="co">#&gt; -  ClonalStatPlot :  ClonalStatPlot     </span></span>
<span><span class="co">#&gt;    Visualize the statistics of the clones.</span></span>
<span><span class="co">#&gt; -  ClonalCompositionPlot :  ClonalCompositionPlot     </span></span>
<span><span class="co">#&gt;    Plot the composition of the clones in different samples/groups.</span></span>
<span><span class="co">#&gt; -  EnrichmentPlot :  Enrichment Plot     </span></span>
<span><span class="co">#&gt;    This function generates various types of plots for enrichment (over-representation) analysis.</span></span>
<span><span class="co">#&gt; -  eq :  Helper functions to select clones based on various criteria     </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; -  CellVelocityPlot :  Cell Velocity Plot     </span></span>
<span><span class="co">#&gt;    This function creates a cell velocity plot for a Seurat object,</span></span>
<span><span class="co">#&gt;    a Giotto object, a path to an .h5ad file or an opened    H5File    by    hdf5r    package.</span></span>
<span><span class="co">#&gt;    It allows for various customizations such as grouping by metadata,</span></span>
<span><span class="co">#&gt;    adding edges between cell neighbors, highlighting specific cells, and more.</span></span>
<span><span class="co">#&gt;    This function is a wrapper around    plotthis::VelocityPlot()   , which provides a</span></span>
<span><span class="co">#&gt;    flexible way to visualize cell velocities in reduced dimensions. This function</span></span>
<span><span class="co">#&gt;    extracts the cell embeddings and velocity embeddings from the Seurat or Giotto object</span></span>
<span><span class="co">#&gt;    and passes them to    plotthis::VelocityPlot()   .</span></span>
<span><span class="co">#&gt; -  SpatFeaturePlot :  Plot features for spatial data     </span></span>
<span><span class="co">#&gt;    The features can include  expression, dimension reduction components, metadata, etc</span></span>
<span><span class="co">#&gt; -  shared :  Helper functions to select clones based on various criteria     </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; -  CellStatPlot :  Cell statistics plot     </span></span>
<span><span class="co">#&gt;    This function creates a plot to visualize the statistics of cells in a Seurat object, a Giotto object,</span></span>
<span><span class="co">#&gt;    a path to an .h5ad file or an opened    H5File    by    hdf5r    package.</span></span>
<span><span class="co">#&gt;    It can create various types of plots, including bar plots, circos plots, pie charts, pies (heatmap with cell_type = 'pie'), ring/donut plots, trend plots</span></span>
<span><span class="co">#&gt;    area plots, sankey/alluvial plots, heatmaps, radar plots, spider plots, violin plots, and box plots.</span></span>
<span><span class="co">#&gt;    The function allows for grouping, splitting, and faceting the data based on metadata columns.</span></span>
<span><span class="co">#&gt;    It also supports calculating fractions of cells based on specified groupings.#'</span></span>
<span><span class="co">#&gt; -  ne :  Helper functions to select clones based on various criteria     </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; -  SpatDimPlot :  Plot categories for spatial data     </span></span>
<span><span class="co">#&gt;    Plot categories for spatial data</span></span>
<span><span class="co">#&gt; -  GSEASummaryPlot :  Objects exported from other packages     </span></span>
<span><span class="co">#&gt;    These objects are imported from other packages. Follow the links</span></span>
<span><span class="co">#&gt;    below to see their documentation.</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;         plotthis   GSEAPlot   ,    GSEASummaryPlot   </span></span>
<span><span class="co">#&gt; -  CCCPlot :  Cell-Cell Communication Plot     </span></span>
<span><span class="co">#&gt;    Plot the cell-cell communication.</span></span>
<span><span class="co">#&gt;    See also:</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;     The review:    https://www.sciencedirect.com/science/article/pii/S2452310021000081   </span></span>
<span><span class="co">#&gt;     The LIANA package:    https://liana-py.readthedocs.io/en/latest/notebooks/basic_usage.html#Tileplot   </span></span>
<span><span class="co">#&gt;     The CCPlotR package:    https://github.com/Sarah145/CCPlotR   </span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt; -  ClonalResidencyPlot :  ClonalResidencyPlot     </span></span>
<span><span class="co">#&gt;    Plot the residency of the clones in different samples.</span></span>
<span><span class="co">#&gt; -  ClustreePlot :  Clustree plot     </span></span>
<span><span class="co">#&gt;    This function generates a clustree plot from a data frame or a Seurat object.</span></span>
<span><span class="co">#&gt; -  ClonalKmerPlot :  ClonalKmerPlot     </span></span>
<span><span class="co">#&gt;    Explore the k-mer frequency of CDR3 sequences.</span></span>
<span><span class="co">#&gt; -  ClonalPositionalPlot :  ClonalPositionalPlot     </span></span>
<span><span class="co">#&gt;    Visualize the positional entropy, property or amino acid frequency of CDR3 sequences.</span></span>
<span><span class="co">#&gt; -  uniq :  Helper functions to select clones based on various criteria     </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; -  top :  Helper functions to select clones based on various criteria     </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; -  le :  Helper functions to select clones based on various criteria     </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; -  MarkersPlot :  Visualize Markers     </span></span>
<span><span class="co">#&gt;    Plot markers, typically identified by    Seurat::FindMarkers()    or    Seurat::FindAllMarkers()   .</span></span>
<span><span class="co">#&gt; -  ClonalOverlapPlot :  ClonalOverlapPlot     </span></span>
<span><span class="co">#&gt;    Plot the overlap of the clones in different samples/groups.</span></span>
<span><span class="co">#&gt; -  ge :  Helper functions to select clones based on various criteria     </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; -  ClonalLengthPlot :  ClonalLengthPlot     </span></span>
<span><span class="co">#&gt;    Plot the length distribution of the CDR3 sequences</span></span>
<span><span class="co">#&gt; -  and :  Helper functions to select clones based on various criteria     </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; -  ClonalAbundancePlot :  ClonalAbundancePlot     </span></span>
<span><span class="co">#&gt;    Plot the count or density of the clones at different abundance levels.</span></span>
<span><span class="co">#&gt; -  sel :  Helper functions to select clones based on various criteria     </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; -  ClonalDiversityPlot :  ClonalDiversityPlot     </span></span>
<span><span class="co">#&gt;    Plot the clonal diversities of the samples/groups.</span></span>
<span><span class="co">#&gt; -  FeatureStatPlot :  Feature statistic plot     </span></span>
<span><span class="co">#&gt;    This function creates various types of feature statistic plots for a Seurat object, a Giotto object,</span></span>
<span><span class="co">#&gt;    a path to an .h5ad file or an opened    H5File    by    hdf5r    package.</span></span>
<span><span class="co">#&gt;    It allows for plotting features such as gene expression, scores, or other metadata across different groups or conditions.</span></span>
<span><span class="co">#&gt;    The function supports multiple plot types including violin, box, bar, ridge, dimension reduction, correlation, heatmap, and dot plots.</span></span>
<span><span class="co">#&gt;    It can also handle multiple features and supports faceting, splitting, and grouping by metadata columns.</span></span>
<span><span class="co">#&gt; -  ClonalGeneUsagePlot :  ClonalGeneUsagePlot     </span></span>
<span><span class="co">#&gt;    ClonalGeneUsagePlot</span></span>
<span><span class="co">#&gt; -  lt :  Helper functions to select clones based on various criteria     </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; -  GSEAPlot :  Objects exported from other packages     </span></span>
<span><span class="co">#&gt;    These objects are imported from other packages. Follow the links</span></span>
<span><span class="co">#&gt;    below to see their documentation.</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;         plotthis   GSEAPlot   ,    GSEASummaryPlot   </span></span>
<span><span class="co">#&gt; -  or :  Helper functions to select clones based on various criteria     </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; -  ListTools :  List all available tools</span></span>
<span><span class="co">#&gt;    List all available tools that can be used to handle the chat request.</span></span>
<span><span class="co">#&gt; -  ListData :  List all available data objects</span></span>
<span><span class="co">#&gt;    List all available data objects that can be used to handle the chat request.</span></span>
<span><span class="co"># or you can ask the LLM to list the available functions</span></span>
<span><span class="co"># chat$ask("List the available functions for visualizing data.")</span></span></code></pre></div>
<p>The tool used for the visualization is determined by the LLM
automatically from your prompt.</p>
</div>
<div class="section level2">
<h2 id="visualize-the-data">Visualize the data<a class="anchor" aria-label="anchor" href="#visualize-the-data"></a>
</h2>
<p>You can visualize the data by using the <code>ask()</code> method.
The LLM will automatically detect the data and the function to be used
for visualization.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chat</span><span class="op">$</span><span class="fu">ask</span><span class="op">(</span><span class="st">"Generate a cell-cell communication plot for the cellphonedb_res data."</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Tool identified:  CCCPlot </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Data object identified:  scplotter::cellphonedb_res</span></span>
<span><span class="co">#&gt; Warning in wrap$modify_fn(prompt_text, llm_provider): The 'skimr' package is</span></span>
<span><span class="co">#&gt; required to skim dataframes. Skim summary of dataframes currently not shown in</span></span>
<span><span class="co">#&gt; prompt</span></span>
<span><span class="co">#&gt; Code ran:</span></span>
<span><span class="co">#&gt; CCCPlot(data = cellphonedb_res, plot_type = "dot")</span></span></code></pre></div>
<p><img src="Visualizing_data_with_LLMs_files/figure-html/unnamed-chunk-6-1.png" class="r-plt" alt="" width="672"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Previous conversation is memorized</span></span>
<span><span class="va">chat</span><span class="op">$</span><span class="fu">ask</span><span class="op">(</span><span class="st">"Do a heatmap instead"</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Tool identified:  CCCPlot </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Data object identified:  scplotter::cellphonedb_res</span></span>
<span><span class="co">#&gt; Warning in wrap$modify_fn(prompt_text, llm_provider): The 'skimr' package is</span></span>
<span><span class="co">#&gt; required to skim dataframes. Skim summary of dataframes currently not shown in</span></span>
<span><span class="co">#&gt; prompt</span></span>
<span><span class="co">#&gt; Code ran:</span></span>
<span><span class="co">#&gt; CCCPlot(cellphonedb_res, plot_type = "heatmap")</span></span></code></pre></div>
<p><img src="Visualizing_data_with_LLMs_files/figure-html/unnamed-chunk-7-1.png" class="r-plt" alt="" width="672"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chat</span><span class="op">$</span><span class="fu">ask</span><span class="op">(</span><span class="st">"Add a proper title to the plot"</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Tool identified:  CCCPlot </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Data object identified:  scplotter::cellphonedb_res</span></span>
<span><span class="co">#&gt; Warning in wrap$modify_fn(prompt_text, llm_provider): The 'skimr' package is</span></span>
<span><span class="co">#&gt; required to skim dataframes. Skim summary of dataframes currently not shown in</span></span>
<span><span class="co">#&gt; prompt</span></span>
<span><span class="co">#&gt; Code ran:</span></span>
<span><span class="co">#&gt; CCCPlot(cellphonedb_res, plot_type = "heatmap", title = "Cell-Cell Communication Heatmap – LIANA-based interactions")</span></span></code></pre></div>
<p><img src="Visualizing_data_with_LLMs_files/figure-html/unnamed-chunk-8-1.png" class="r-plt" alt="" width="672"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># To fetch the previous conversation</span></span>
<span><span class="co"># Note that the response from the LLM is simplified in the history</span></span>
<span><span class="va">chat</span><span class="op">$</span><span class="fu">get_history</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "User: Generate a cell-cell communication plot for the cellphonedb_res data."                                                                                                                   </span></span>
<span><span class="co">#&gt; [2] "Assistant: tool - CCCPlot; data - scplotter::cellphonedb_res; code - CCCPlot(data = cellphonedb_res, plot_type = \"dot\")"                                                                     </span></span>
<span><span class="co">#&gt; [3] "User: Do a heatmap instead"                                                                                                                                                                    </span></span>
<span><span class="co">#&gt; [4] "Assistant: tool - CCCPlot; data - scplotter::cellphonedb_res; code - CCCPlot(cellphonedb_res, plot_type = \"heatmap\")"                                                                        </span></span>
<span><span class="co">#&gt; [5] "User: Add a proper title to the plot"                                                                                                                                                          </span></span>
<span><span class="co">#&gt; [6] "Assistant: tool - CCCPlot; data - scplotter::cellphonedb_res; code - CCCPlot(cellphonedb_res, plot_type = \"heatmap\", title = \"Cell-Cell Communication Heatmap – LIANA-based interactions\")"</span></span>
<span></span>
<span><span class="co"># To clear the history</span></span>
<span><span class="va">chat</span><span class="op">$</span><span class="fu">clear_history</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="debug-and-improve-the-prompt">Debug and improve the prompt<a class="anchor" aria-label="anchor" href="#debug-and-improve-the-prompt"></a>
</h2>
<p>You can set <code>verbose</code> to <code>TRUE</code> for all
conversations when constructing the <code>chat</code> object. This will
print the prompt and the response from the LLM.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chat</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/SCPlotterChat.html">SCPlotterChat</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span></span>
<span>    provider <span class="op">=</span> <span class="va">provider</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">chat</span><span class="op">$</span><span class="fu">ask</span><span class="op">(</span><span class="st">"Generate a cell-cell communication plot for the cellphonedb_res data."</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">--- Sending request to LLM provider (gpt-5-nano): ---</span></span></span>
<span><span class="co">#&gt; Objective: Select the most appropriate tool to handle the user's request while preserving conversational context. If the user refines or changes how the previous result should be visualized (e.g., asks for a different plot type), continue with the last plotting tool used unless they explicitly name a different tool.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Decision Process:</span></span>
<span><span class="co">#&gt; - If the user explicitly names a tool, output that tool.</span></span>
<span><span class="co">#&gt; - Else, if the request appears to refine the previous output (e.g., "do X instead", "make it a heatmap/dot/bar/etc", "change to ...", "add ... to", "same plot but ..."), select the last tool mentioned in the chat history.</span></span>
<span><span class="co">#&gt; - Else, analyze the current user request; if there is a clear, unambiguous match to a single tool, select that tool.</span></span>
<span><span class="co">#&gt; - Else, use the last mentioned tool from the chat history.</span></span>
<span><span class="co">#&gt; - If no tool is found, respond with "None".</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Response Format: Provide only the name of the selected tool, or "None" if no tool applies.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; User Request:</span></span>
<span><span class="co">#&gt; Generate a cell-cell communication plot for the cellphonedb_res data.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Available Tools:</span></span>
<span><span class="co">#&gt; - gt: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - CellDimPlot: Cell Dimension Reduction Plot  </span></span>
<span><span class="co">#&gt;    This function creates a dimension reduction plot for a Seurat object</span></span>
<span><span class="co">#&gt;    a Giotto object, a path to an .h5ad file or an opened    H5File    by    hdf5r    package.</span></span>
<span><span class="co">#&gt;    It allows for various customizations such as grouping by metadata,</span></span>
<span><span class="co">#&gt;    adding edges between cell neighbors, highlighting specific cells, and more.</span></span>
<span><span class="co">#&gt;    This function is a wrapper around    plotthis::DimPlot()   , which provides a</span></span>
<span><span class="co">#&gt;    flexible way to visualize cell clusters in reduced dimensions. This function</span></span>
<span><span class="co">#&gt;    extracts the necessary data from the Seurat or Giotto object and passes it to</span></span>
<span><span class="co">#&gt;    plotthis::DimPlot()   .</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalVolumePlot: ClonalVolumePlot  </span></span>
<span><span class="co">#&gt;    ClonalVolumePlot</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalRarefactionPlot: ClonalRarefactionPlot  </span></span>
<span><span class="co">#&gt;    Plot the rarefaction curves</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalStatPlot: ClonalStatPlot  </span></span>
<span><span class="co">#&gt;    Visualize the statistics of the clones.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalCompositionPlot: ClonalCompositionPlot  </span></span>
<span><span class="co">#&gt;    Plot the composition of the clones in different samples/groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - EnrichmentPlot: Enrichment Plot  </span></span>
<span><span class="co">#&gt;    This function generates various types of plots for enrichment (over-representation) analysis.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - eq: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - CellVelocityPlot: Cell Velocity Plot  </span></span>
<span><span class="co">#&gt;    This function creates a cell velocity plot for a Seurat object,</span></span>
<span><span class="co">#&gt;    a Giotto object, a path to an .h5ad file or an opened    H5File    by    hdf5r    package.</span></span>
<span><span class="co">#&gt;    It allows for various customizations such as grouping by metadata,</span></span>
<span><span class="co">#&gt;    adding edges between cell neighbors, highlighting specific cells, and more.</span></span>
<span><span class="co">#&gt;    This function is a wrapper around    plotthis::VelocityPlot()   , which provides a</span></span>
<span><span class="co">#&gt;    flexible way to visualize cell velocities in reduced dimensions. This function</span></span>
<span><span class="co">#&gt;    extracts the cell embeddings and velocity embeddings from the Seurat or Giotto object</span></span>
<span><span class="co">#&gt;    and passes them to    plotthis::VelocityPlot()   .</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - SpatFeaturePlot: Plot features for spatial data  </span></span>
<span><span class="co">#&gt;    The features can include  expression, dimension reduction components, metadata, etc</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - shared: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - CellStatPlot: Cell statistics plot  </span></span>
<span><span class="co">#&gt;    This function creates a plot to visualize the statistics of cells in a Seurat object, a Giotto object,</span></span>
<span><span class="co">#&gt;    a path to an .h5ad file or an opened    H5File    by    hdf5r    package.</span></span>
<span><span class="co">#&gt;    It can create various types of plots, including bar plots, circos plots, pie charts, pies (heatmap with cell_type = 'pie'), ring/donut plots, trend plots</span></span>
<span><span class="co">#&gt;    area plots, sankey/alluvial plots, heatmaps, radar plots, spider plots, violin plots, and box plots.</span></span>
<span><span class="co">#&gt;    The function allows for grouping, splitting, and faceting the data based on metadata columns.</span></span>
<span><span class="co">#&gt;    It also supports calculating fractions of cells based on specified groupings.#'</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ne: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - SpatDimPlot: Plot categories for spatial data  </span></span>
<span><span class="co">#&gt;    Plot categories for spatial data</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - GSEASummaryPlot: Objects exported from other packages  </span></span>
<span><span class="co">#&gt;    These objects are imported from other packages. Follow the links</span></span>
<span><span class="co">#&gt;    below to see their documentation.</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;         plotthis   GSEAPlot   ,    GSEASummaryPlot   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - CCCPlot: Cell-Cell Communication Plot  </span></span>
<span><span class="co">#&gt;    Plot the cell-cell communication.</span></span>
<span><span class="co">#&gt;    See also:</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;     The review:    https://www.sciencedirect.com/science/article/pii/S2452310021000081   </span></span>
<span><span class="co">#&gt;     The LIANA package:    https://liana-py.readthedocs.io/en/latest/notebooks/basic_usage.html#Tileplot   </span></span>
<span><span class="co">#&gt;     The CCPlotR package:    https://github.com/Sarah145/CCPlotR   </span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalResidencyPlot: ClonalResidencyPlot  </span></span>
<span><span class="co">#&gt;    Plot the residency of the clones in different samples.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClustreePlot: Clustree plot  </span></span>
<span><span class="co">#&gt;    This function generates a clustree plot from a data frame or a Seurat object.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalKmerPlot: ClonalKmerPlot  </span></span>
<span><span class="co">#&gt;    Explore the k-mer frequency of CDR3 sequences.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalPositionalPlot: ClonalPositionalPlot  </span></span>
<span><span class="co">#&gt;    Visualize the positional entropy, property or amino acid frequency of CDR3 sequences.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - uniq: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - top: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - le: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - MarkersPlot: Visualize Markers  </span></span>
<span><span class="co">#&gt;    Plot markers, typically identified by    Seurat::FindMarkers()    or    Seurat::FindAllMarkers()   .</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalOverlapPlot: ClonalOverlapPlot  </span></span>
<span><span class="co">#&gt;    Plot the overlap of the clones in different samples/groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ge: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalLengthPlot: ClonalLengthPlot  </span></span>
<span><span class="co">#&gt;    Plot the length distribution of the CDR3 sequences</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - and: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalAbundancePlot: ClonalAbundancePlot  </span></span>
<span><span class="co">#&gt;    Plot the count or density of the clones at different abundance levels.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - sel: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalDiversityPlot: ClonalDiversityPlot  </span></span>
<span><span class="co">#&gt;    Plot the clonal diversities of the samples/groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - FeatureStatPlot: Feature statistic plot  </span></span>
<span><span class="co">#&gt;    This function creates various types of feature statistic plots for a Seurat object, a Giotto object,</span></span>
<span><span class="co">#&gt;    a path to an .h5ad file or an opened    H5File    by    hdf5r    package.</span></span>
<span><span class="co">#&gt;    It allows for plotting features such as gene expression, scores, or other metadata across different groups or conditions.</span></span>
<span><span class="co">#&gt;    The function supports multiple plot types including violin, box, bar, ridge, dimension reduction, correlation, heatmap, and dot plots.</span></span>
<span><span class="co">#&gt;    It can also handle multiple features and supports faceting, splitting, and grouping by metadata columns.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalGeneUsagePlot: ClonalGeneUsagePlot  </span></span>
<span><span class="co">#&gt;    ClonalGeneUsagePlot</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - lt: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - GSEAPlot: Objects exported from other packages  </span></span>
<span><span class="co">#&gt;    These objects are imported from other packages. Follow the links</span></span>
<span><span class="co">#&gt;    below to see their documentation.</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;         plotthis   GSEAPlot   ,    GSEASummaryPlot   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - or: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ListTools: List all available tools</span></span>
<span><span class="co">#&gt; List all available tools that can be used to handle the chat request.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ListData: List all available data objects</span></span>
<span><span class="co">#&gt; List all available data objects that can be used to handle the chat request.</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">--- Receiving response from LLM provider: ---</span></span></span>
<span><span class="co">#&gt; CCCPlot</span></span>
<span><span class="co">#&gt; Tool identified:  CCCPlot</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">--- Sending request to LLM provider (gpt-5-nano): ---</span></span></span>
<span><span class="co">#&gt; Objective: Identify the data object to be used.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Decision Process:</span></span>
<span><span class="co">#&gt; - If the user explicitly names a data object, output that data object name. If the name does not match exactly the available ones, use the one that is mostly related.</span></span>
<span><span class="co">#&gt; - Else, if the request appears to refine the previous output (e.g., "do X instead", "make it a heatmap/dot/bar/etc", "change to ...", "add ... to", "same plot but ..."), select the last dataset name mentioned in the chat history.</span></span>
<span><span class="co">#&gt; - Else, analyze the current user request; if there is a clear, unambiguous match to a dataset, select that data object.</span></span>
<span><span class="co">#&gt; - Else, use the last mentioned data object from the chat history.</span></span>
<span><span class="co">#&gt; - If no data object is found, respond with "None".</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Response Format: Provide only the name of the selected data object, or "None" if no tool applies. The response should be unquoted.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; User Request:</span></span>
<span><span class="co">#&gt; Generate a cell-cell communication plot for the cellphonedb_res data.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Available Data Objects:</span></span>
<span><span class="co">#&gt; - scplotter::cellphonedb_res: A toy example of CellPhoneDB output from LIANA</span></span>
<span><span class="co">#&gt; - scplotter::ifnb_sub: A subsetted version of 'ifnb' datasets</span></span>
<span><span class="co">#&gt; - scplotter::pancreas_sub: A subsetted version of mouse 'pancreas' datasets</span></span>
<span><span class="co">#&gt; - Seurat::cc.genes: Cell cycle genes</span></span>
<span><span class="co">#&gt; - Seurat::cc.genes.updated.2019: Cell cycle genes: 2019 update</span></span>
<span><span class="co">#&gt; - SeuratObject::pbmc_small: A small example version of the PBMC dataset</span></span>
<span><span class="co">#&gt; - scRepertoire::contig_list: A list of 8 single-cell T cell receptor sequences runs.</span></span>
<span><span class="co">#&gt; - scRepertoire::mini_contig_list: Processed subset of 'contig_list'</span></span>
<span><span class="co">#&gt; - scRepertoire::scRep_example: A Seurat object of 500 single T cells,</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">--- Receiving response from LLM provider: ---</span></span></span>
<span><span class="co">#&gt; scplotter::cellphonedb_res</span></span>
<span><span class="co">#&gt; Data object identified:  scplotter::cellphonedb_res</span></span>
<span><span class="co">#&gt; Warning in wrap$modify_fn(prompt_text, llm_provider): The 'skimr' package is</span></span>
<span><span class="co">#&gt; required to skim dataframes. Skim summary of dataframes currently not shown in</span></span>
<span><span class="co">#&gt; prompt</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">--- Sending request to LLM provider (gpt-5-nano): ---</span></span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Objective: Generate the R code to run the specified tool using the specified data object based on the user's request and the provided tool information.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Decision Process:</span></span>
<span><span class="co">#&gt; - Analyze the user's request to understand what needs to be done with the specified tool and data object.</span></span>
<span><span class="co">#&gt; - When use the data object name, do not quote it.</span></span>
<span><span class="co">#&gt; - Refer to the provided tool information to understand the tool's usage, arguments, and examples.</span></span>
<span><span class="co">#&gt; - Consider any specific parameters or options mentioned in the user's request that need to be included in the tool call.</span></span>
<span><span class="co">#&gt; - If the user's request is ambiguous or lacks necessary details, refer to the chat history for additional context that may help clarify the intended use of the tool and data object.</span></span>
<span><span class="co">#&gt; - Construct the R code that correctly calls the tool with the specified data object, ensuring that all necessary arguments are included and correctly formatted.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Response Format: Provide only the valid R code wrapped between "```r" and "```" to run the tool.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Tool to be used: CCCPlot</span></span>
<span><span class="co">#&gt; Data object to be used: cellphonedb_res</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; User Request:</span></span>
<span><span class="co">#&gt; Generate a cell-cell communication plot for the cellphonedb_res data.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Tool Information:</span></span>
<span><span class="co">#&gt; - title</span></span>
<span><span class="co">#&gt;   Cell-Cell Communication Plot</span></span>
<span><span class="co">#&gt; - description</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;    Plot the cell-cell communication.</span></span>
<span><span class="co">#&gt;    See also:</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;     The review:  https://www.sciencedirect.com/science/article/pii/S2452310021000081 </span></span>
<span><span class="co">#&gt;     The LIANA package:  https://liana-py.readthedocs.io/en/latest/notebooks/basic_usage.html#Tileplot </span></span>
<span><span class="co">#&gt;     The CCPlotR package:  https://github.com/Sarah145/CCPlotR </span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt; - usage</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;   CCCPlot(</span></span>
<span><span class="co">#&gt;     data,</span></span>
<span><span class="co">#&gt;     plot_type = c("dot", "network", "chord", "circos", "heatmap", "sankey", "alluvial",</span></span>
<span><span class="co">#&gt;       "box", "violin", "ridge"),</span></span>
<span><span class="co">#&gt;     method = c("aggregation", "interaction"),</span></span>
<span><span class="co">#&gt;     magnitude = waiver(),</span></span>
<span><span class="co">#&gt;     specificity = waiver(),</span></span>
<span><span class="co">#&gt;     magnitude_agg = length,</span></span>
<span><span class="co">#&gt;     magnitude_name = "No. of interactions",</span></span>
<span><span class="co">#&gt;     meta_specificity = "sumlog",</span></span>
<span><span class="co">#&gt;     split_by = NULL,</span></span>
<span><span class="co">#&gt;     x_text_angle = 90,</span></span>
<span><span class="co">#&gt;     link_curvature = 0.2,</span></span>
<span><span class="co">#&gt;     link_alpha = 0.6,</span></span>
<span><span class="co">#&gt;     facet_by = NULL,</span></span>
<span><span class="co">#&gt;     show_row_names = TRUE,</span></span>
<span><span class="co">#&gt;     show_column_names = TRUE,</span></span>
<span><span class="co">#&gt;     ...</span></span>
<span><span class="co">#&gt;   )</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt; - arguments</span></span>
<span><span class="co">#&gt;   - data: A data frame with the cell-cell communication data.</span></span>
<span><span class="co">#&gt;   A typical data frame should have the following columns:</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;    source The source cell type.</span></span>
<span><span class="co">#&gt;    target The target cell type.</span></span>
<span><span class="co">#&gt;    ligand The ligand gene.</span></span>
<span><span class="co">#&gt;    receptor The receptor gene.</span></span>
<span><span class="co">#&gt;    ligand_means The mean expression of the ligand gene per cell type.</span></span>
<span><span class="co">#&gt;    receptor_means The mean expression of the receptor gene per cell type.</span></span>
<span><span class="co">#&gt;    ligand_props The proportion of cells that express the entity.</span></span>
<span><span class="co">#&gt;    receptor_props The proportion of cells that express the entity.</span></span>
<span><span class="co">#&gt;    &lt;magnitude&gt; The magnitude of the communication.</span></span>
<span><span class="co">#&gt;    &lt;specificity&gt; The specificity of the communication.</span></span>
<span><span class="co">#&gt;   Depends on the plot_type, some columns are optional. But the source, target,</span></span>
<span><span class="co">#&gt;   ligand, receptor and &lt;magnitude&gt; are required.</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;   - plot_type: The type of plot to use. Default is "dot".</span></span>
<span><span class="co">#&gt;   Possible values are "network", "chord", "circos", "heatmap", "sankey", "alluvial", "dot",</span></span>
<span><span class="co">#&gt;   "box", "violin" and "ridge".</span></span>
<span><span class="co">#&gt;   For "box", "violin" and "ridge", the method should be "interaction".</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;    network: A network plot with the source and target cells as the nodes and the communication as the edges.</span></span>
<span><span class="co">#&gt;    chord: A chord plot with the source and target cells as the nodes and the communication as the chords.</span></span>
<span><span class="co">#&gt;    circos: Alias of "chord".</span></span>
<span><span class="co">#&gt;    heatmap: A heatmap plot with the source and target cells as the rows and columns.</span></span>
<span><span class="co">#&gt;    sankey: A sankey plot with the source and target cells as the nodes and the communication as the flows.</span></span>
<span><span class="co">#&gt;    alluvial: Alias of "sankey".</span></span>
<span><span class="co">#&gt;    dot: A dot plot with the source and target cells as the nodes and the communication as the dots.</span></span>
<span><span class="co">#&gt;    box: Box plots for source cell types. Each x is a target cell type and the values will be</span></span>
<span><span class="co">#&gt;   the interaction strengths of the ligand-receptor pairs.</span></span>
<span><span class="co">#&gt;    violin: Violin plots for source cell types. Each x is a target cell type and the values will be</span></span>
<span><span class="co">#&gt;   the interaction strengths of the ligand-receptor pairs.</span></span>
<span><span class="co">#&gt;    ridge: Ridge plots for source cell types. Each row is a target cell type and the values will be</span></span>
<span><span class="co">#&gt;   the interaction strengths of the ligand-receptor pairs.</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;   - method: The method to determine the plot entities.</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;    aggregation: Aggregate the ligand-receptor pairs interactions for each source-target pair.</span></span>
<span><span class="co">#&gt;   Only the source / target pairs will be plotted.</span></span>
<span><span class="co">#&gt;    interaction: Plot the ligand-receptor pairs interactions directly.</span></span>
<span><span class="co">#&gt;   The ligand-receptor pairs will also be plotted.</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;   - magnitude: The column name in the data to use as the magnitude of the communication.</span></span>
<span><span class="co">#&gt;   By default, the second last column will be used.</span></span>
<span><span class="co">#&gt;   See li.mt.show_methods() for the available methods in LIANA.</span></span>
<span><span class="co">#&gt;   or https://liana-py.readthedocs.io/en/latest/notebooks/basic_usage.html#Tileplot</span></span>
<span><span class="co">#&gt;   - specificity: The column name in the data to use as the specificity of the communication.</span></span>
<span><span class="co">#&gt;   By default, the last column will be used.</span></span>
<span><span class="co">#&gt;   If the method doesn't have a specificity, set it to NULL.</span></span>
<span><span class="co">#&gt;   - magnitude_agg: A function to aggregate the magnitude of the communication.</span></span>
<span><span class="co">#&gt;   Default is length.</span></span>
<span><span class="co">#&gt;   - magnitude_name: The name of the magnitude in the plot.</span></span>
<span><span class="co">#&gt;   Default is "No. of interactions".</span></span>
<span><span class="co">#&gt;   - meta_specificity: The method to calculate the specificity when there are multiple</span></span>
<span><span class="co">#&gt;   ligand-receptor pairs interactions. Default is "sumlog".</span></span>
<span><span class="co">#&gt;   It should be one of the methods in the metap package.</span></span>
<span><span class="co">#&gt;   Current available methods are:</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;    invchisq: Combine p values using the inverse chi squared method</span></span>
<span><span class="co">#&gt;    invt: Combine p values using the inverse t method</span></span>
<span><span class="co">#&gt;    logitp: Combine p values using the logit method</span></span>
<span><span class="co">#&gt;    meanp: Combine p values by the mean p method</span></span>
<span><span class="co">#&gt;    meanz: Combine p values using the mean z method</span></span>
<span><span class="co">#&gt;    sumlog: Combine p-values by the sum of logs (Fisher's) method</span></span>
<span><span class="co">#&gt;    sump: Combine p-values using the sum of p (Edgington's) method</span></span>
<span><span class="co">#&gt;    two2one: Convert two-sided p-values to one-sided</span></span>
<span><span class="co">#&gt;    votep: Combine p-values by the vote counting method</span></span>
<span><span class="co">#&gt;    wilkinsonp: Combine p-values using Wilkinson's method</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;   - split_by: A character vector of column names to split the plots. Default is NULL.</span></span>
<span><span class="co">#&gt;   - x_text_angle: The angle of the x-axis text. Default is 90.</span></span>
<span><span class="co">#&gt;   Only used when plot_type is "dot".</span></span>
<span><span class="co">#&gt;   - link_curvature: The curvature of the links. Default is 0.2.</span></span>
<span><span class="co">#&gt;   Only used when plot_type is "network".</span></span>
<span><span class="co">#&gt;   - link_alpha: The transparency of the links. Default is 0.6.</span></span>
<span><span class="co">#&gt;   Only used when plot_type is "network".</span></span>
<span><span class="co">#&gt;   - facet_by: A character vector of column names to facet the plots. Default is NULL.</span></span>
<span><span class="co">#&gt;   It should always be NULL.</span></span>
<span><span class="co">#&gt;   - show_row_names: Whether to show the row names in the heatmap. Default is TRUE.</span></span>
<span><span class="co">#&gt;   Only used when plot_type is "heatmap".</span></span>
<span><span class="co">#&gt;   - show_column_names: Whether to show the column names in the heatmap. Default is TRUE.</span></span>
<span><span class="co">#&gt;   Only used when plot_type is "heatmap".</span></span>
<span><span class="co">#&gt;   - ...: Other arguments passed to the specific plot function.</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;    For Network, see plotthis::Network().</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;       [...] can be:</span></span>
<span><span class="co">#&gt;       - links: A data frame containing the links between nodes.</span></span>
<span><span class="co">#&gt;       - nodes: A data frame containing the nodes.</span></span>
<span><span class="co">#&gt;       This is optional. The names of the nodes are extracted from the links data frame.</span></span>
<span><span class="co">#&gt;       If "@nodes" is provided, the nodes data frame will be extracted from the attribute nodes of the links data frame.</span></span>
<span><span class="co">#&gt;       - split_by_sep: The separator for multiple split_by columns. See split_by</span></span>
<span><span class="co">#&gt;       - split_nodes: A logical value specifying whether to split the nodes data.</span></span>
<span><span class="co">#&gt;       If TRUE, the nodes data will also be split by the split_by column.</span></span>
<span><span class="co">#&gt;       - from: A character string specifying the column name of the links data frame for the source nodes.</span></span>
<span><span class="co">#&gt;       Default is the first column of the links data frame.</span></span>
<span><span class="co">#&gt;       - from_sep: A character string to concatenate the columns in from, if multiple columns are provided.</span></span>
<span><span class="co">#&gt;       - to: A character string specifying the column name of the links data frame for the target nodes.</span></span>
<span><span class="co">#&gt;       Default is the second column of the links data frame.</span></span>
<span><span class="co">#&gt;       - to_sep: A character string to concatenate the columns in to, if multiple columns are provided.</span></span>
<span><span class="co">#&gt;       - node_by: A character string specifying the column name of the nodes data frame for the node names.</span></span>
<span><span class="co">#&gt;       Default is the first column of the nodes data frame.</span></span>
<span><span class="co">#&gt;       - node_by_sep: A character string to concatenate the columns in node_by, if multiple columns are provided.</span></span>
<span><span class="co">#&gt;       - link_weight_by: A numeric value or a character string specifying the column name of the links data frame for the link weight.</span></span>
<span><span class="co">#&gt;       If a numeric value is provided, all links will have the same weight.</span></span>
<span><span class="co">#&gt;       This determines the width of the links.</span></span>
<span><span class="co">#&gt;       - link_weight_name: A character string specifying the name of the link weight in the legend.</span></span>
<span><span class="co">#&gt;       - link_type_by: A character string specifying the type of the links.</span></span>
<span><span class="co">#&gt;       This can be "solid", "dashed", "dotted", or a column name from the links data frame.</span></span>
<span><span class="co">#&gt;       It has higher priority when it is a column name.</span></span>
<span><span class="co">#&gt;       - link_type_name: A character string specifying the name of the link type in the legend.</span></span>
<span><span class="co">#&gt;       - node_size_by: A numeric value or a character string specifying the column name of the nodes data frame for the node size.</span></span>
<span><span class="co">#&gt;       If a numeric value is provided, all nodes will have the same size.</span></span>
<span><span class="co">#&gt;       - node_size_name: A character string specifying the name of the node size in the legend.</span></span>
<span><span class="co">#&gt;       - node_color_by: A character string specifying the color of the nodes.</span></span>
<span><span class="co">#&gt;       This can be a color name, a hex code, or a column name from the nodes data frame.</span></span>
<span><span class="co">#&gt;       It has higher priority when it is a column name.</span></span>
<span><span class="co">#&gt;       - node_color_name: A character string specifying the name of the node color in the legend.</span></span>
<span><span class="co">#&gt;       - node_shape_by: A numeric value or a character string specifying the column name of the nodes data frame for the node shape.</span></span>
<span><span class="co">#&gt;       If a numeric value is provided, all nodes will have the same shape.</span></span>
<span><span class="co">#&gt;       - node_shape_name: A character string specifying the name of the node shape in the legend.</span></span>
<span><span class="co">#&gt;       - node_fill_by: A character string specifying the fill color of the nodes.</span></span>
<span><span class="co">#&gt;       This can be a color name, a hex code, or a column name from the nodes data frame.</span></span>
<span><span class="co">#&gt;       It has higher priority when it is a column name.</span></span>
<span><span class="co">#&gt;       - node_fill_name: A character string specifying the name of the node fill in the legend.</span></span>
<span><span class="co">#&gt;       - node_alpha: A numeric value specifying the transparency of the nodes.</span></span>
<span><span class="co">#&gt;       It only works when the nodes are filled.</span></span>
<span><span class="co">#&gt;       - node_stroke: A numeric value specifying the stroke of the nodes.</span></span>
<span><span class="co">#&gt;       - cluster_scale: A character string specifying how to scale the clusters.</span></span>
<span><span class="co">#&gt;       It can be "fill", "color", or "shape".</span></span>
<span><span class="co">#&gt;       - node_size_range: A numeric vector specifying the range of the node size.</span></span>
<span><span class="co">#&gt;       - link_weight_range: A numeric vector specifying the range of the link weight.</span></span>
<span><span class="co">#&gt;       - link_arrow_offset: A numeric value specifying the offset of the link arrows.</span></span>
<span><span class="co">#&gt;       So that they won't overlap with the nodes.</span></span>
<span><span class="co">#&gt;       - link_color_by: A character string specifying the colors of the link. It can be:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        "from" means the color of the link is determined by the source node.</span></span>
<span><span class="co">#&gt;        "to" means the color of the link is determined by the target node.</span></span>
<span><span class="co">#&gt;        Otherwise, the color of the link is determined by the column name from the links data frame.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - link_color_name: A character string specifying the name of the link color in the legend.</span></span>
<span><span class="co">#&gt;       Only used when link_color_by is a column name.</span></span>
<span><span class="co">#&gt;       - palette: A character string specifying the palette to use.</span></span>
<span><span class="co">#&gt;       A named list or vector can be used to specify the palettes for different split_by values.</span></span>
<span><span class="co">#&gt;       - palcolor: A character string specifying the color to use in the palette.</span></span>
<span><span class="co">#&gt;       A named list can be used to specify the colors for different split_by values.</span></span>
<span><span class="co">#&gt;       If some values are missing, the values from the palette will be used (palcolor will be NULL for those values).</span></span>
<span><span class="co">#&gt;       - link_palette: A character string specifying the palette of the links.</span></span>
<span><span class="co">#&gt;       When link_color_by is "from" or "to", the palette of the links defaults to the palette of the nodes.</span></span>
<span><span class="co">#&gt;       - link_palcolor: A character vector specifying the colors of the link palette.</span></span>
<span><span class="co">#&gt;       When link_color_by is "from" or "to", the colors of the link palette defaults to the colors of the node palette.</span></span>
<span><span class="co">#&gt;       - directed: A logical value specifying whether the graph is directed.</span></span>
<span><span class="co">#&gt;       - layout: A character string specifying the layout of the graph.</span></span>
<span><span class="co">#&gt;       It can be "circle", "tree", "grid", or a layout function from igraph.</span></span>
<span><span class="co">#&gt;       - cluster: A character string specifying the clustering method.</span></span>
<span><span class="co">#&gt;       It can be "none", "fast_greedy", "walktrap", "edge_betweenness", "infomap", or a clustering function from igraph.</span></span>
<span><span class="co">#&gt;       - add_mark: A logical value specifying whether to add mark for the clusters to the plot.</span></span>
<span><span class="co">#&gt;       - mark_expand: A unit value specifying the expansion of the mark.</span></span>
<span><span class="co">#&gt;       - mark_type: A character string specifying the type of the mark.</span></span>
<span><span class="co">#&gt;       It can be "hull", "ellipse", "rect", "circle", or a mark function from ggforce.</span></span>
<span><span class="co">#&gt;       - mark_alpha: A numeric value specifying the transparency of the mark.</span></span>
<span><span class="co">#&gt;       - mark_linetype: A numeric value specifying the line type of the mark.</span></span>
<span><span class="co">#&gt;       - add_label: A logical value specifying whether to add label to the nodes to the plot.</span></span>
<span><span class="co">#&gt;       - label_size: A numeric value specifying the size of the label.</span></span>
<span><span class="co">#&gt;       - label_fg: A character string specifying the foreground color of the label.</span></span>
<span><span class="co">#&gt;       - label_bg: A character string specifying the background color of the label.</span></span>
<span><span class="co">#&gt;       - label_bg_r: A numeric value specifying the background ratio of the label.</span></span>
<span><span class="co">#&gt;       - arrow: An arrow object for the links.</span></span>
<span><span class="co">#&gt;       - title: A character string specifying the title of the plot.</span></span>
<span><span class="co">#&gt;       A function can be used to generate the title based on the default title.</span></span>
<span><span class="co">#&gt;       This is useful when split_by is used and the title needs to be dynamic.</span></span>
<span><span class="co">#&gt;       - subtitle: A character string specifying the subtitle of the plot.</span></span>
<span><span class="co">#&gt;       - xlab: A character string specifying the x-axis label.</span></span>
<span><span class="co">#&gt;       - ylab: A character string specifying the y-axis label.</span></span>
<span><span class="co">#&gt;       - aspect.ratio: A numeric value specifying the aspect ratio of the plot.</span></span>
<span><span class="co">#&gt;       - theme: A character string or a theme class (i.e. ggplot2::theme_classic) specifying the theme to use.</span></span>
<span><span class="co">#&gt;       Default is "theme_this".</span></span>
<span><span class="co">#&gt;       - theme_args: A list of arguments to pass to the theme function.</span></span>
<span><span class="co">#&gt;       - legend.position: A character string specifying the position of the legend.</span></span>
<span><span class="co">#&gt;       if waiver(), for single groups, the legend will be "none", otherwise "right".</span></span>
<span><span class="co">#&gt;       - legend.direction: A character string specifying the direction of the legend.</span></span>
<span><span class="co">#&gt;       - seed: The random seed to use. Default is 8525.</span></span>
<span><span class="co">#&gt;       - combine: Whether to combine the plots into one when facet is FALSE. Default is TRUE.</span></span>
<span><span class="co">#&gt;       - nrow: A numeric value specifying the number of rows in the facet.</span></span>
<span><span class="co">#&gt;       - ncol: A numeric value specifying the number of columns in the facet.</span></span>
<span><span class="co">#&gt;       - byrow: A logical value indicating whether to fill the plots by row.</span></span>
<span><span class="co">#&gt;       - axes: A string specifying how axes should be treated. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'keep' will retain all axes in individual plots.</span></span>
<span><span class="co">#&gt;        'collect' will remove duplicated axes when placed in the same run of rows or columns of the layout.</span></span>
<span><span class="co">#&gt;        'collect_x' and 'collect_y' will remove duplicated x-axes in the columns or duplicated y-axes in the rows respectively.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - axis_titles: A string specifying how axis titltes should be treated. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'keep' will retain all axis titles in individual plots.</span></span>
<span><span class="co">#&gt;        'collect' will remove duplicated titles in one direction and merge titles in the opposite direction.</span></span>
<span><span class="co">#&gt;        'collect_x' and 'collect_y' control this for x-axis titles and y-axis titles respectively.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - guides: A string specifying how guides should be treated in the layout. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'collect' will collect guides below to the given nesting level, removing duplicates.</span></span>
<span><span class="co">#&gt;        'keep' will stop collection at this level and let guides be placed alongside their plot.</span></span>
<span><span class="co">#&gt;        'auto' will allow guides to be collected if a upper level tries, but place them alongside the plot if not.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - design: Specification of the location of areas in the layout, passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE. When specified, nrow, ncol, and byrow are ignored.</span></span>
<span><span class="co">#&gt;       See patchwork::wrap_plots() for more details.</span></span>
<span><span class="co">#&gt;   For ChordPlot, see plotthis::ChordPlot().</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;       [...] can be:</span></span>
<span><span class="co">#&gt;       - y: A character string specifying the column name of the data frame to plot for the y-axis.</span></span>
<span><span class="co">#&gt;       - from: A character string of the column name to plot for the source.</span></span>
<span><span class="co">#&gt;       A character/factor column is expected.</span></span>
<span><span class="co">#&gt;       - from_sep: A character string to concatenate the columns in from, if multiple columns are provided.</span></span>
<span><span class="co">#&gt;       - to: A character string of the column name to plot for the target.</span></span>
<span><span class="co">#&gt;       A character/factor column is expected.</span></span>
<span><span class="co">#&gt;       - to_sep: A character string to concatenate the columns in to, if multiple columns are provided.</span></span>
<span><span class="co">#&gt;       - split_by_sep: The separator for multiple split_by columns. See split_by</span></span>
<span><span class="co">#&gt;       - flip: A logical value to flip the source and target.</span></span>
<span><span class="co">#&gt;       - links_color: A character string to specify the color of the links.</span></span>
<span><span class="co">#&gt;       Either "from" or "to".</span></span>
<span><span class="co">#&gt;       - theme: A character string or a theme class (i.e. ggplot2::theme_classic) specifying the theme to use.</span></span>
<span><span class="co">#&gt;       Default is "theme_this".</span></span>
<span><span class="co">#&gt;       - theme_args: A list of arguments to pass to the theme function.</span></span>
<span><span class="co">#&gt;       - palette: A character string specifying the palette to use.</span></span>
<span><span class="co">#&gt;       A named list or vector can be used to specify the palettes for different split_by values.</span></span>
<span><span class="co">#&gt;       - palcolor: A character string specifying the color to use in the palette.</span></span>
<span><span class="co">#&gt;       A named list can be used to specify the colors for different split_by values.</span></span>
<span><span class="co">#&gt;       If some values are missing, the values from the palette will be used (palcolor will be NULL for those values).</span></span>
<span><span class="co">#&gt;       - alpha: A numeric value specifying the transparency of the plot.</span></span>
<span><span class="co">#&gt;       - labels_rot: A logical value to rotate the labels by 90 degrees.</span></span>
<span><span class="co">#&gt;       - title: A character string specifying the title of the plot.</span></span>
<span><span class="co">#&gt;       A function can be used to generate the title based on the default title.</span></span>
<span><span class="co">#&gt;       This is useful when split_by is used and the title needs to be dynamic.</span></span>
<span><span class="co">#&gt;       - subtitle: A character string specifying the subtitle of the plot.</span></span>
<span><span class="co">#&gt;       - seed: The random seed to use. Default is 8525.</span></span>
<span><span class="co">#&gt;       - keep_na: A logical value or a character to replace the NA values in the data.</span></span>
<span><span class="co">#&gt;       It can also take a named list to specify different behavior for different columns.</span></span>
<span><span class="co">#&gt;       If TRUE or NA, NA values will be replaced with NA.</span></span>
<span><span class="co">#&gt;       If FALSE, NA values will be removed from the data before plotting.</span></span>
<span><span class="co">#&gt;       If a character string is provided, NA values will be replaced with the provided string.</span></span>
<span><span class="co">#&gt;       If a named vector/list is provided, the names should be the column names to apply the behavior to,</span></span>
<span><span class="co">#&gt;       and the values should be one of TRUE, FALSE, or a character string.</span></span>
<span><span class="co">#&gt;       Without a named vector/list, the behavior applies to categorical/character columns used on the plot,</span></span>
<span><span class="co">#&gt;       for example, the x, group_by, fill_by, etc.</span></span>
<span><span class="co">#&gt;       - keep_empty: One of FALSE, TRUE and "level". It can also take a named list to specify</span></span>
<span><span class="co">#&gt;       different behavior for different columns. Without a named list, the behavior applies to the</span></span>
<span><span class="co">#&gt;       categorical/character columns used on the plot, for example, the x, group_by, fill_by, etc.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       FALSE (default): Drop empty factor levels from the data before plotting.</span></span>
<span><span class="co">#&gt;       TRUE: Keep empty factor levels and show them as a separate category in the plot.</span></span>
<span><span class="co">#&gt;       "level": Keep empty factor levels, but do not show them in the plot.</span></span>
<span><span class="co">#&gt;       But they will be assigned colors from the palette to maintain consistency across multiple plots.</span></span>
<span><span class="co">#&gt;       Alias: levels</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - combine: Whether to combine the plots into one when facet is FALSE. Default is TRUE.</span></span>
<span><span class="co">#&gt;       - nrow: A numeric value specifying the number of rows in the facet.</span></span>
<span><span class="co">#&gt;       - ncol: A numeric value specifying the number of columns in the facet.</span></span>
<span><span class="co">#&gt;       - byrow: A logical value indicating whether to fill the plots by row.</span></span>
<span><span class="co">#&gt;       - axes: A string specifying how axes should be treated. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'keep' will retain all axes in individual plots.</span></span>
<span><span class="co">#&gt;        'collect' will remove duplicated axes when placed in the same run of rows or columns of the layout.</span></span>
<span><span class="co">#&gt;        'collect_x' and 'collect_y' will remove duplicated x-axes in the columns or duplicated y-axes in the rows respectively.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - axis_titles: A string specifying how axis titltes should be treated. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'keep' will retain all axis titles in individual plots.</span></span>
<span><span class="co">#&gt;        'collect' will remove duplicated titles in one direction and merge titles in the opposite direction.</span></span>
<span><span class="co">#&gt;        'collect_x' and 'collect_y' control this for x-axis titles and y-axis titles respectively.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - guides: A string specifying how guides should be treated in the layout. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'collect' will collect guides below to the given nesting level, removing duplicates.</span></span>
<span><span class="co">#&gt;        'keep' will stop collection at this level and let guides be placed alongside their plot.</span></span>
<span><span class="co">#&gt;        'auto' will allow guides to be collected if a upper level tries, but place them alongside the plot if not.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - design: Specification of the location of areas in the layout, passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE. When specified, nrow, ncol, and byrow are ignored.</span></span>
<span><span class="co">#&gt;       See patchwork::wrap_plots() for more details.</span></span>
<span><span class="co">#&gt;   For Heatmap, see plotthis::Heatmap().</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;       [...] can be:</span></span>
<span><span class="co">#&gt;       - values_by: A character of column name in data that contains the values to be plotted.</span></span>
<span><span class="co">#&gt;       This is required when in_form is "long". For other formats, the values are pivoted into a column named by values_by.</span></span>
<span><span class="co">#&gt;       - values_fill: A value to fill in the missing values in the heatmap.</span></span>
<span><span class="co">#&gt;       When there is missing value in the data, the cluster_rows and cluster_columns will fail.</span></span>
<span><span class="co">#&gt;       - name: A character string to name the heatmap (will be used to rename values_by).</span></span>
<span><span class="co">#&gt;       - in_form: The format of the data. Can be one of "matrix", "long", "wide-rows", "wide-columns", or "auto".</span></span>
<span><span class="co">#&gt;       Defaults to "auto".</span></span>
<span><span class="co">#&gt;       - split_by_sep: A character string to concat multiple columns in split_by.</span></span>
<span><span class="co">#&gt;       - rows_by: A vector of column names in data that contains the row information.</span></span>
<span><span class="co">#&gt;       This is used to create the rows of the heatmap.</span></span>
<span><span class="co">#&gt;       When in_form is "long" or "wide-columns", this is requied, and multiple columns can be specified,</span></span>
<span><span class="co">#&gt;       which will be concatenated by rows_by_sep into a single column.</span></span>
<span><span class="co">#&gt;       - rows_by_sep: A character string to concat multiple columns in rows_by.</span></span>
<span><span class="co">#&gt;       - rows_split_by: A character of column name in data that contains the split information for rows.</span></span>
<span><span class="co">#&gt;       - rows_split_by_sep: A character string to concat multiple columns in rows_split_by.</span></span>
<span><span class="co">#&gt;       - columns_by: A vector of column names in data that contains the column information.</span></span>
<span><span class="co">#&gt;       This is used to create the columns of the heatmap.</span></span>
<span><span class="co">#&gt;       When in_form is "long" or "wide-rows", this is required, and multiple columns can be specified,</span></span>
<span><span class="co">#&gt;       which will be concatenated by columns_by_sep into a single column.</span></span>
<span><span class="co">#&gt;       - columns_by_sep: A character string to concat multiple columns in columns_by.</span></span>
<span><span class="co">#&gt;       - columns_split_by: A character of column name in data that contains the split information for columns.</span></span>
<span><span class="co">#&gt;       - columns_split_by_sep: A character string to concat multiple columns in columns_split_by.</span></span>
<span><span class="co">#&gt;       - rows_data: A data frame containing additional data for rows, which can be used to add annotations to the heatmap.</span></span>
<span><span class="co">#&gt;       It will be joined to the main data by rows_by and split_by if split_by exists in rows_data.</span></span>
<span><span class="co">#&gt;       This is useful for adding additional information to the rows of the heatmap.</span></span>
<span><span class="co">#&gt;       - columns_data: A data frame containing additional data for columns, which can be used to add annotations to the heatmap.</span></span>
<span><span class="co">#&gt;       It will be joined to the main data by columns_by and split_by if split_by exists in columns_data.</span></span>
<span><span class="co">#&gt;       This is useful for adding additional information to the columns of the heatmap.</span></span>
<span><span class="co">#&gt;       - keep_na: Whether we should keep NA groups in rows, columns and split_by variables. Default is FALSE.</span></span>
<span><span class="co">#&gt;       FALSE to remove NA groups; TRUE to keep NA groups.</span></span>
<span><span class="co">#&gt;       A vector of column names can also be provided to specify which columns to keep NA groups.</span></span>
<span><span class="co">#&gt;       Note that the record will be removed if any of the grouping columns has NA and is not specified to keep NA.</span></span>
<span><span class="co">#&gt;       - keep_empty: One of FALSE, TRUE and "level". It can also take a named list to specify</span></span>
<span><span class="co">#&gt;       different behavior for different columns. Without a named list, the behavior applies to the</span></span>
<span><span class="co">#&gt;       categorical/character columns used on the plot, for example, the x, group_by, fill_by, etc.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       FALSE (default): Drop empty factor levels from the data before plotting.</span></span>
<span><span class="co">#&gt;       TRUE: Keep empty factor levels and show them as a separate category in the plot.</span></span>
<span><span class="co">#&gt;       "level": Keep empty factor levels, but do not show them in the plot.</span></span>
<span><span class="co">#&gt;       But they will be assigned colors from the palette to maintain consistency across multiple plots.</span></span>
<span><span class="co">#&gt;       Alias: levels</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - rows_orderby: A expression (in character) to specify how to order rows. It will be evaluated in the context of the data frame used for rows (after grouping by rows_split_by and rows_by). The expression should return a vector of the same length as the number of rows in the data frame. The default is NULL, which means no specific ordering.</span></span>
<span><span class="co">#&gt;       Can't be used with cluster_rows = TRUE.</span></span>
<span><span class="co">#&gt;       This is applied before renaming rows_by to rows_name.</span></span>
<span><span class="co">#&gt;       - columns_orderby: A expression (in character) to specify how to order columns. It will be evaluated in the context of the data frame used for columns (after grouping by columns</span></span>
<span><span class="co">#&gt;       split_by and columns_by). The expression should return a vector of the same length as the number of rows in the data frame. The default is NULL, which means no specific ordering.</span></span>
<span><span class="co">#&gt;       Can't be used with cluster_columns = TRUE.</span></span>
<span><span class="co">#&gt;       This is applied before renaming columns_by to columns_name.</span></span>
<span><span class="co">#&gt;       - columns_name: A character string to rename the column created by columns_by, which will be reflected in the name of the annotation or legend.</span></span>
<span><span class="co">#&gt;       - columns_split_name: A character string to rename the column created by columns_split_by, which will be reflected in the name of the annotation or legend.</span></span>
<span><span class="co">#&gt;       - rows_name: A character string to rename the column created by rows_by, which will be reflected in the name of the annotation or legend.</span></span>
<span><span class="co">#&gt;       - rows_split_name: A character string to rename the column created by rows_split_by, which will be reflected in the name of the annotation or legend.</span></span>
<span><span class="co">#&gt;       - palette: A character string specifying the palette of the heatmap cells.</span></span>
<span><span class="co">#&gt;       - palcolor: A character vector of colors to override the palette of the heatmap cells.</span></span>
<span><span class="co">#&gt;       - rows_palette: A character string specifying the palette of the row group annotation.</span></span>
<span><span class="co">#&gt;       The default is "Paired".</span></span>
<span><span class="co">#&gt;       - rows_palcolor: A character vector of colors to override the palette of the row group annotation.</span></span>
<span><span class="co">#&gt;       - rows_split_palette: A character string specifying the palette of the row split annotation.</span></span>
<span><span class="co">#&gt;       The default is "simspec".</span></span>
<span><span class="co">#&gt;       - rows_split_palcolor: A character vector of colors to override the palette of the row split annotation.</span></span>
<span><span class="co">#&gt;       - columns_palette: A character string specifying the palette of the column group annotation.</span></span>
<span><span class="co">#&gt;       The default is "Paired".</span></span>
<span><span class="co">#&gt;       - columns_palcolor: A character vector of colors to override the palette of the column group annotation.</span></span>
<span><span class="co">#&gt;       - columns_split_palette: A character string specifying the palette of the column split annotation.</span></span>
<span><span class="co">#&gt;       The default is "simspec".</span></span>
<span><span class="co">#&gt;       - columns_split_palcolor: A character vector of colors to override the palette of the column split annotation.</span></span>
<span><span class="co">#&gt;       - pie_size_name: A character string specifying the name of the legend for the pie size.</span></span>
<span><span class="co">#&gt;       - pie_size: A numeric value or a function specifying the size of the pie chart.</span></span>
<span><span class="co">#&gt;       If it is a function, the function should take count as the argument and return the size.</span></span>
<span><span class="co">#&gt;       - pie_values: A function or character that can be converted to a function by match.arg()</span></span>
<span><span class="co">#&gt;       to calculate the values for the pie chart. Default is "length".</span></span>
<span><span class="co">#&gt;       The function should take a vector of values as the argument and return a single value, for each</span></span>
<span><span class="co">#&gt;       group in pie_group_by.</span></span>
<span><span class="co">#&gt;       - pie_name: A character string to rename the column created by pie_group_by, which will be reflected in the name of the annotation or legend.</span></span>
<span><span class="co">#&gt;       - pie_group_by: A character of column name in data that contains the group information for pie charts.</span></span>
<span><span class="co">#&gt;       This is used to create pie charts in the heatmap when cell_type is "pie".</span></span>
<span><span class="co">#&gt;       - pie_group_by_sep: A character string to concat multiple columns in pie_group_by.</span></span>
<span><span class="co">#&gt;       - pie_palette: A character string specifying the palette of the pie chart.</span></span>
<span><span class="co">#&gt;       - pie_palcolor: A character vector of colors to override the palette of the pie chart.</span></span>
<span><span class="co">#&gt;       - bars_sample: An integer specifying the number of samples to draw the bars.</span></span>
<span><span class="co">#&gt;       - label: A function to calculate the labels for the heatmap cells.</span></span>
<span><span class="co">#&gt;       It can take either 1, 3, or 5 arguments. The first argument is the aggregated values.</span></span>
<span><span class="co">#&gt;       If it takes 3 arguments, the second and third arguments are the row and column indices.</span></span>
<span><span class="co">#&gt;       If it takes 5 arguments, the second and third arguments are the row and column indices,</span></span>
<span><span class="co">#&gt;       the fourth and fifth arguments are the row and column names.</span></span>
<span><span class="co">#&gt;       The function should return a character vector of the same length as the aggregated values.</span></span>
<span><span class="co">#&gt;       If the function returns NA, no label will be shown for that cell.</span></span>
<span><span class="co">#&gt;       For the indices, if you have the same dimension of data (same order of rows and columns) as the heatmap, you need to use ComplexHeatmap::pindex() to get the correct values.</span></span>
<span><span class="co">#&gt;       - label_size: A numeric value specifying the size of the labels when cell_type = "label".</span></span>
<span><span class="co">#&gt;       - violin_fill: A character vector of colors to override the fill color of the violin plot.</span></span>
<span><span class="co">#&gt;       If NULL, the fill color will be the same as the annotion.</span></span>
<span><span class="co">#&gt;       - boxplot_fill: A character vector of colors to override the fill color of the boxplot.</span></span>
<span><span class="co">#&gt;       If NULL, the fill color will be the same as the annotion.</span></span>
<span><span class="co">#&gt;       - dot_size: A numeric value specifying the size of the dot or a function to calculate the size</span></span>
<span><span class="co">#&gt;       from the values in the cell or a function to calculate the size from the values in the cell.</span></span>
<span><span class="co">#&gt;       - dot_size_name: A character string specifying the name of the legend for the dot size.</span></span>
<span><span class="co">#&gt;       If NULL, the dot size legend will not be shown.</span></span>
<span><span class="co">#&gt;       - legend_items: A numeric vector with names to specifiy the items in the main legend.</span></span>
<span><span class="co">#&gt;       The names will be working as the labels of the legend items.</span></span>
<span><span class="co">#&gt;       - legend_discrete: A logical value indicating whether the main legend is discrete.</span></span>
<span><span class="co">#&gt;       - legend.position: A character string specifying the position of the legend.</span></span>
<span><span class="co">#&gt;       if waiver(), for single groups, the legend will be "none", otherwise "right".</span></span>
<span><span class="co">#&gt;       - legend.direction: A character string specifying the direction of the legend.</span></span>
<span><span class="co">#&gt;       - lower_quantile, upper_quantile, lower_cutoff, upper_cutoff: Vector of minimum and maximum cutoff values or quantile values for each feature.</span></span>
<span><span class="co">#&gt;       It's applied to aggregated values when aggregated values are used (e.g. plot_type tile, label, etc).</span></span>
<span><span class="co">#&gt;       It's applied to raw values when raw values are used (e.g. plot_type bars, etc).</span></span>
<span><span class="co">#&gt;       - add_bg: A logical value indicating whether to add a background to the heatmap.</span></span>
<span><span class="co">#&gt;       Does not work with cell_type = "bars" or cell_type = "tile".</span></span>
<span><span class="co">#&gt;       - bg_alpha: A numeric value between 0 and 1 specifying the transparency of the background.</span></span>
<span><span class="co">#&gt;       - add_reticle: A logical value indicating whether to add a reticle to the heatmap.</span></span>
<span><span class="co">#&gt;       - reticle_color: A character string specifying the color of the reticle.</span></span>
<span><span class="co">#&gt;       - column_name_annotation: A logical value indicating whether to add the column annotation for the column names.</span></span>
<span><span class="co">#&gt;       which is a simple annotaion indicating the column names.</span></span>
<span><span class="co">#&gt;       - column_name_legend: A logical value indicating whether to show the legend of the column name annotation.</span></span>
<span><span class="co">#&gt;       - row_name_annotation: A logical value indicating whether to add the row annotation for the row names.</span></span>
<span><span class="co">#&gt;       which is a simple annotaion indicating the row names.</span></span>
<span><span class="co">#&gt;       - row_name_legend: A logical value indicating whether to show the legend of the row name annotation.</span></span>
<span><span class="co">#&gt;       - cluster_columns: A logical value indicating whether to cluster the columns.</span></span>
<span><span class="co">#&gt;       If TRUE and columns_split_by is provided, the clustering will only be applied to the columns within the same split.</span></span>
<span><span class="co">#&gt;       - cluster_rows: A logical value indicating whether to cluster the rows.</span></span>
<span><span class="co">#&gt;       If TRUE and rows_split_by is provided, the clustering will only be applied to the rows within the same split.</span></span>
<span><span class="co">#&gt;       - border: A logical value indicating whether to draw the border of the heatmap.</span></span>
<span><span class="co">#&gt;       If TRUE, the borders of the slices will be also drawn.</span></span>
<span><span class="co">#&gt;       - title: The global (column) title of the heatmap</span></span>
<span><span class="co">#&gt;       - column_title: A character string/vector of the column name(s) to use as the title of the column group annotation.</span></span>
<span><span class="co">#&gt;       - row_title: A character string/vector of the column name(s) to use as the title of the row group annotation.</span></span>
<span><span class="co">#&gt;       - na_col: A character string specifying the color for missing values.</span></span>
<span><span class="co">#&gt;       The default is "grey85".</span></span>
<span><span class="co">#&gt;       - row_names_side: A character string specifying the side of the row names.</span></span>
<span><span class="co">#&gt;       The default is "right".</span></span>
<span><span class="co">#&gt;       - column_names_side: A character string specifying the side of the column names.</span></span>
<span><span class="co">#&gt;       The default is "bottom".</span></span>
<span><span class="co">#&gt;       - column_annotation: A character string/vector of the column name(s) to use as the column annotation.</span></span>
<span><span class="co">#&gt;       Or a list with the keys as the names of the annotation and the values as the column names.</span></span>
<span><span class="co">#&gt;       - column_annotation_side: A character string specifying the side of the column annotation.</span></span>
<span><span class="co">#&gt;       Could be a list with the keys as the names of the annotation and the values as the sides.</span></span>
<span><span class="co">#&gt;       - column_annotation_palette: A character string specifying the palette of the column annotation.</span></span>
<span><span class="co">#&gt;       The default is "Paired".</span></span>
<span><span class="co">#&gt;       Could be a list with the keys as the names of the annotation and the values as the palettes.</span></span>
<span><span class="co">#&gt;       - column_annotation_palcolor: A character vector of colors to override the palette of the column annotation.</span></span>
<span><span class="co">#&gt;       Could be a list with the keys as the names of the annotation and the values as the palcolors.</span></span>
<span><span class="co">#&gt;       - column_annotation_type: A character string specifying the type of the column annotation.</span></span>
<span><span class="co">#&gt;       The default is "auto". Other options are "simple", "pie", "ring", "bar", "violin", "boxplot", "density".</span></span>
<span><span class="co">#&gt;       Could be a list with the keys as the names of the annotation and the values as the types.</span></span>
<span><span class="co">#&gt;       If the type is "auto", the type will be determined by the type and number of the column data.</span></span>
<span><span class="co">#&gt;       - column_annotation_params: A list of parameters passed to the annotation function.</span></span>
<span><span class="co">#&gt;       Could be a list with the keys as the names of the annotation and the values as the parameters passed to the annotation function. For the parameters for names (columns_by, rows_by, columns_split_by, rows_split_by), the key should be "name.(name)", where (name) is the name of the annotation.</span></span>
<span><span class="co">#&gt;       See anno_pie(), anno_ring(), anno_bar(), anno_violin(), anno_boxplot(), anno_density(), anno_simple(), anno_points() and anno_lines() for the parameters of each annotation function.</span></span>
<span><span class="co">#&gt;       - column_annotation_agg: A function to aggregate the values in the column annotation.</span></span>
<span><span class="co">#&gt;       - row_annotation: A character string/vector of the column name(s) to use as the row annotation.</span></span>
<span><span class="co">#&gt;       Or a list with the keys as the names of the annotation and the values as the column names.</span></span>
<span><span class="co">#&gt;       - row_annotation_side: A character string specifying the side of the row annotation.</span></span>
<span><span class="co">#&gt;       Could be a list with the keys as the names of the annotation and the values as the sides.</span></span>
<span><span class="co">#&gt;       - row_annotation_palette: A character string specifying the palette of the row annotation.</span></span>
<span><span class="co">#&gt;       The default is "Paired".</span></span>
<span><span class="co">#&gt;       Could be a list with the keys as the names of the annotation and the values as the palettes.</span></span>
<span><span class="co">#&gt;       - row_annotation_palcolor: A character vector of colors to override the palette of the row annotation.</span></span>
<span><span class="co">#&gt;       Could be a list with the keys as the names of the annotation and the values as the palcolors.</span></span>
<span><span class="co">#&gt;       - row_annotation_type: A character string specifying the type of the row annotation.</span></span>
<span><span class="co">#&gt;       The default is "auto". Other options are "simple", "pie", "ring", "bar", "violin", "boxplot", "density".</span></span>
<span><span class="co">#&gt;       Could be a list with the keys as the names of the annotation and the values as the types.</span></span>
<span><span class="co">#&gt;       If the type is "auto", the type will be determined by the type and number of the row data.</span></span>
<span><span class="co">#&gt;       - row_annotation_params: A list of parameters passed to the annotation function.</span></span>
<span><span class="co">#&gt;       Could be a list with the keys as the names of the annotation and the values as the parameters.</span></span>
<span><span class="co">#&gt;       Same as column_annotation_params.</span></span>
<span><span class="co">#&gt;       - row_annotation_agg: A function to aggregate the values in the row annotation.</span></span>
<span><span class="co">#&gt;       - flip: A logical value indicating whether to flip the heatmap.</span></span>
<span><span class="co">#&gt;       The idea is that, you can simply set flip = TRUE to flip the heatmap.</span></span>
<span><span class="co">#&gt;       You don't need to swap the arguments related to rows and columns, except those you specify via ...</span></span>
<span><span class="co">#&gt;       that are passed to ComplexHeatmap::Heatmap() directly.</span></span>
<span><span class="co">#&gt;       - alpha: A numeric value between 0 and 1 specifying the transparency of the heatmap cells.</span></span>
<span><span class="co">#&gt;       - seed: The random seed to use. Default is 8525.</span></span>
<span><span class="co">#&gt;       - layer_fun_callback: A function to add additional layers to the heatmap.</span></span>
<span><span class="co">#&gt;       The function should have the following arguments: j, i, x, y, w, h, fill, sr and sc.</span></span>
<span><span class="co">#&gt;       Please also refer to the layer_fun argument in ComplexHeatmap::Heatmap.</span></span>
<span><span class="co">#&gt;       - cell_type: A character string specifying the type of the heatmap cells.</span></span>
<span><span class="co">#&gt;       The default is values. Other options are "bars", "label", "dot", "violin", "boxplot".</span></span>
<span><span class="co">#&gt;       Note that for pie chart, the values under columns specified by rows will not be used directly. Instead, the values</span></span>
<span><span class="co">#&gt;       will just be counted in different pie_group_by groups. NA values will not be counted.</span></span>
<span><span class="co">#&gt;       - cell_agg: A function to aggregate the values in the cell, for the cell type "tile" and "label".</span></span>
<span><span class="co">#&gt;       The default is mean.</span></span>
<span><span class="co">#&gt;       - combine: Whether to combine the plots into one when facet is FALSE. Default is TRUE.</span></span>
<span><span class="co">#&gt;       - nrow: A numeric value specifying the number of rows in the facet.</span></span>
<span><span class="co">#&gt;       - ncol: A numeric value specifying the number of columns in the facet.</span></span>
<span><span class="co">#&gt;       - byrow: A logical value indicating whether to fill the plots by row.</span></span>
<span><span class="co">#&gt;       - axes: A string specifying how axes should be treated. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'keep' will retain all axes in individual plots.</span></span>
<span><span class="co">#&gt;        'collect' will remove duplicated axes when placed in the same run of rows or columns of the layout.</span></span>
<span><span class="co">#&gt;        'collect_x' and 'collect_y' will remove duplicated x-axes in the columns or duplicated y-axes in the rows respectively.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - axis_titles: A string specifying how axis titltes should be treated. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'keep' will retain all axis titles in individual plots.</span></span>
<span><span class="co">#&gt;        'collect' will remove duplicated titles in one direction and merge titles in the opposite direction.</span></span>
<span><span class="co">#&gt;        'collect_x' and 'collect_y' control this for x-axis titles and y-axis titles respectively.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - guides: A string specifying how guides should be treated in the layout. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'collect' will collect guides below to the given nesting level, removing duplicates.</span></span>
<span><span class="co">#&gt;        'keep' will stop collection at this level and let guides be placed alongside their plot.</span></span>
<span><span class="co">#&gt;        'auto' will allow guides to be collected if a upper level tries, but place them alongside the plot if not.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - design: Specification of the location of areas in the layout, passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE. When specified, nrow, ncol, and byrow are ignored.</span></span>
<span><span class="co">#&gt;       See patchwork::wrap_plots() for more details.</span></span>
<span><span class="co">#&gt;   For SankeyPlot, see plotthis::SankeyPlot().</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;       [...] can be:</span></span>
<span><span class="co">#&gt;       - in_form: A character string to specify the format of the data.</span></span>
<span><span class="co">#&gt;       Possible values are "auto", "long", "lodes", "wide", "alluvia", and "counts".</span></span>
<span><span class="co">#&gt;       - x: A character string specifying the column name of the data frame to plot for the x-axis.</span></span>
<span><span class="co">#&gt;       - x_sep: A character string to concatenate the columns in x, if multiple columns are provided.</span></span>
<span><span class="co">#&gt;       - y: A character string specifying the column name of the data frame to plot for the y-axis.</span></span>
<span><span class="co">#&gt;       - stratum: A character string of the column name to group the nodes for each x.</span></span>
<span><span class="co">#&gt;       See data for more details.</span></span>
<span><span class="co">#&gt;       - stratum_sep: A character string to concatenate the columns in stratum, if multiple columns are provided.</span></span>
<span><span class="co">#&gt;       - alluvium: A character string of the column name to define the links.</span></span>
<span><span class="co">#&gt;       See data for more details.</span></span>
<span><span class="co">#&gt;       - alluvium_sep: A character string to concatenate the columns in alluvium, if multiple columns are provided.</span></span>
<span><span class="co">#&gt;       - split_by_sep: The separator for multiple split_by columns. See split_by</span></span>
<span><span class="co">#&gt;       - keep_empty: One of FALSE, TRUE and "level". It can also take a named list to specify</span></span>
<span><span class="co">#&gt;       different behavior for different columns. Without a named list, the behavior applies to the</span></span>
<span><span class="co">#&gt;       categorical/character columns used on the plot, for example, the x, group_by, fill_by, etc.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       FALSE (default): Drop empty factor levels from the data before plotting.</span></span>
<span><span class="co">#&gt;       TRUE: Keep empty factor levels and show them as a separate category in the plot.</span></span>
<span><span class="co">#&gt;       "level": Keep empty factor levels, but do not show them in the plot.</span></span>
<span><span class="co">#&gt;       But they will be assigned colors from the palette to maintain consistency across multiple plots.</span></span>
<span><span class="co">#&gt;       Alias: levels</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - flow: A logical value to use ggalluvial::geom_flow instead of ggalluvial::geom_alluvium.</span></span>
<span><span class="co">#&gt;       - expand: The values to expand the x and y axes. It is like CSS padding.</span></span>
<span><span class="co">#&gt;       When a single value is provided, it is used for both axes on both sides.</span></span>
<span><span class="co">#&gt;       When two values are provided, the first value is used for the top/bottom side and the second value is used for the left/right side.</span></span>
<span><span class="co">#&gt;       When three values are provided, the first value is used for the top side, the second value is used for the left/right side, and the third value is used for the bottom side.</span></span>
<span><span class="co">#&gt;       When four values are provided, the values are used for the top, right, bottom, and left sides, respectively.</span></span>
<span><span class="co">#&gt;       You can also use a named vector to specify the values for each side.</span></span>
<span><span class="co">#&gt;       When the axis is discrete, the values will be applied as 'add' to the 'expansion' function.</span></span>
<span><span class="co">#&gt;       When the axis is continuous, the values will be applied as 'mult' to the 'expansion' function.</span></span>
<span><span class="co">#&gt;       See also https://ggplot2.tidyverse.org/reference/expansion.html</span></span>
<span><span class="co">#&gt;       - nodes_legend: Controls how the legend of nodes will be shown. Possible values are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        "merge": Merge the legends of nodes. That is only one legend will be shown for all nodes.</span></span>
<span><span class="co">#&gt;        "separate": Show the legends of nodes separately. That is, nodes on each x will have their own legend.</span></span>
<span><span class="co">#&gt;        "none": Do not show the legend of nodes.</span></span>
<span><span class="co">#&gt;        "auto": Automatically determine how to show the legend.</span></span>
<span><span class="co">#&gt;       When nodes_label is TRUE, "none" will apply.</span></span>
<span><span class="co">#&gt;       When nodes_label is FALSE, and if stratum is the same as links_fill_by, "none" will apply.</span></span>
<span><span class="co">#&gt;       If there is any overlapping values between the nodes on different x,</span></span>
<span><span class="co">#&gt;       "merge" will apply. Otherwise, "separate" will apply.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - nodes_color: A character string to color the nodes.</span></span>
<span><span class="co">#&gt;       Use a special value ".fill" to use the same color as the fill.</span></span>
<span><span class="co">#&gt;       - links_fill_by: A character string of the column name to fill the links.</span></span>
<span><span class="co">#&gt;       - links_fill_by_sep: A character string to concatenate the columns in links_fill_by, if multiple columns are provided.</span></span>
<span><span class="co">#&gt;       - links_name: A character string to name the legend of links.</span></span>
<span><span class="co">#&gt;       - links_color: A character string to color the borders of links.</span></span>
<span><span class="co">#&gt;       Use a special value ".fill" to use the same color as the fill.</span></span>
<span><span class="co">#&gt;       - nodes_palette: A character string to specify the palette of nodes fill.</span></span>
<span><span class="co">#&gt;       - nodes_palcolor: A character vector to specify the colors of nodes fill.</span></span>
<span><span class="co">#&gt;       - nodes_alpha: A numeric value to specify the transparency of nodes fill.</span></span>
<span><span class="co">#&gt;       - nodes_label: A logical value to show the labels on the nodes.</span></span>
<span><span class="co">#&gt;       - nodes_label_miny: A numeric value to specify the minimum y (frequency) to show the labels.</span></span>
<span><span class="co">#&gt;       - nodes_width: A numeric value to specify the width of nodes.</span></span>
<span><span class="co">#&gt;       - links_palette: A character string to specify the palette of links fill.</span></span>
<span><span class="co">#&gt;       - links_palcolor: A character vector to specify the colors of links fill.</span></span>
<span><span class="co">#&gt;       - links_alpha: A numeric value to specify the transparency of links fill.</span></span>
<span><span class="co">#&gt;       - legend.box: A character string to specify the box of the legend, either "vertical" or "horizontal".</span></span>
<span><span class="co">#&gt;       - aspect.ratio: A numeric value specifying the aspect ratio of the plot.</span></span>
<span><span class="co">#&gt;       - legend.position: A character string specifying the position of the legend.</span></span>
<span><span class="co">#&gt;       if waiver(), for single groups, the legend will be "none", otherwise "right".</span></span>
<span><span class="co">#&gt;       - legend.direction: A character string specifying the direction of the legend.</span></span>
<span><span class="co">#&gt;       - flip: A logical value to flip the plot.</span></span>
<span><span class="co">#&gt;       - theme: A character string or a theme class (i.e. ggplot2::theme_classic) specifying the theme to use.</span></span>
<span><span class="co">#&gt;       Default is "theme_this".</span></span>
<span><span class="co">#&gt;       - theme_args: A list of arguments to pass to the theme function.</span></span>
<span><span class="co">#&gt;       - title: A character string specifying the title of the plot.</span></span>
<span><span class="co">#&gt;       A function can be used to generate the title based on the default title.</span></span>
<span><span class="co">#&gt;       This is useful when split_by is used and the title needs to be dynamic.</span></span>
<span><span class="co">#&gt;       - subtitle: A character string specifying the subtitle of the plot.</span></span>
<span><span class="co">#&gt;       - xlab: A character string specifying the x-axis label.</span></span>
<span><span class="co">#&gt;       - ylab: A character string specifying the y-axis label.</span></span>
<span><span class="co">#&gt;       - facet_scales: Whether to scale the axes of facets. Default is "fixed"</span></span>
<span><span class="co">#&gt;       Other options are "free", "free_x", "free_y". See ggplot2::facet_wrap</span></span>
<span><span class="co">#&gt;       - facet_ncol: A numeric value specifying the number of columns in the facet.</span></span>
<span><span class="co">#&gt;       When facet_by is a single column and facet_wrap is used.</span></span>
<span><span class="co">#&gt;       - facet_nrow: A numeric value specifying the number of rows in the facet.</span></span>
<span><span class="co">#&gt;       When facet_by is a single column and facet_wrap is used.</span></span>
<span><span class="co">#&gt;       - facet_byrow: A logical value indicating whether to fill the plots by row. Default is TRUE.</span></span>
<span><span class="co">#&gt;       - seed: The random seed to use. Default is 8525.</span></span>
<span><span class="co">#&gt;       - combine: Whether to combine the plots into one when facet is FALSE. Default is TRUE.</span></span>
<span><span class="co">#&gt;       - nrow: A numeric value specifying the number of rows in the facet.</span></span>
<span><span class="co">#&gt;       - ncol: A numeric value specifying the number of columns in the facet.</span></span>
<span><span class="co">#&gt;       - byrow: A logical value indicating whether to fill the plots by row.</span></span>
<span><span class="co">#&gt;       - axes: A string specifying how axes should be treated. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'keep' will retain all axes in individual plots.</span></span>
<span><span class="co">#&gt;        'collect' will remove duplicated axes when placed in the same run of rows or columns of the layout.</span></span>
<span><span class="co">#&gt;        'collect_x' and 'collect_y' will remove duplicated x-axes in the columns or duplicated y-axes in the rows respectively.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - axis_titles: A string specifying how axis titltes should be treated. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'keep' will retain all axis titles in individual plots.</span></span>
<span><span class="co">#&gt;        'collect' will remove duplicated titles in one direction and merge titles in the opposite direction.</span></span>
<span><span class="co">#&gt;        'collect_x' and 'collect_y' control this for x-axis titles and y-axis titles respectively.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - guides: A string specifying how guides should be treated in the layout. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'collect' will collect guides below to the given nesting level, removing duplicates.</span></span>
<span><span class="co">#&gt;        'keep' will stop collection at this level and let guides be placed alongside their plot.</span></span>
<span><span class="co">#&gt;        'auto' will allow guides to be collected if a upper level tries, but place them alongside the plot if not.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - design: Specification of the location of areas in the layout, passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE. When specified, nrow, ncol, and byrow are ignored.</span></span>
<span><span class="co">#&gt;       See patchwork::wrap_plots() for more details.</span></span>
<span><span class="co">#&gt;   For DotPlot, see plotthis::DotPlot().</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;       [...] can be:</span></span>
<span><span class="co">#&gt;       - x: A character vector specifying the column to use for the x-axis.</span></span>
<span><span class="co">#&gt;       A numeric column is expected.</span></span>
<span><span class="co">#&gt;       - y: A character vector specifying the column to use for the y-axis.</span></span>
<span><span class="co">#&gt;       A factor/character column is expected.</span></span>
<span><span class="co">#&gt;       - x_sep: A character vector to concatenate multiple columns in x. Default is "_".</span></span>
<span><span class="co">#&gt;       - y_sep: A character vector to concatenate multiple columns in y. Default is "_".</span></span>
<span><span class="co">#&gt;       - flip: A logical value indicating whether to flip the x and y axes. Default is FALSE.</span></span>
<span><span class="co">#&gt;       - split_by_sep: The separator for multiple split_by columns. See split_by</span></span>
<span><span class="co">#&gt;       - size_name: A character vector specifying the name for the size legend.</span></span>
<span><span class="co">#&gt;       - fill_name: A character vector specifying the name for the fill legend.</span></span>
<span><span class="co">#&gt;       - fill_cutoff_name: A character vector specifying the name for the fill cutoff legend.</span></span>
<span><span class="co">#&gt;       - add_bg: A logical value indicating whether to add a background color to the plot. Default is FALSE.</span></span>
<span><span class="co">#&gt;       - bg_palette: A character vector specifying the palette for the background color. Default is "stripe".</span></span>
<span><span class="co">#&gt;       - bg_palcolor: A character vector specifying the color for the background color.</span></span>
<span><span class="co">#&gt;       - bg_alpha: A numeric value specifying the alpha for the background color. Default is 0.2.</span></span>
<span><span class="co">#&gt;       - bg_direction: A character vector specifying the direction for the background color. Default is "vertical".</span></span>
<span><span class="co">#&gt;       Other options are "horizontal". "h" and "v" are also accepted.</span></span>
<span><span class="co">#&gt;       - size_by: Which column to use as the size of the dots. It must be a numeric column.</span></span>
<span><span class="co">#&gt;       If not provided, the size will be the count of the instances for each 'y' in 'x'.</span></span>
<span><span class="co">#&gt;       For 'ScatterPlot', it can be a single numeric value to specify the size of the dots.</span></span>
<span><span class="co">#&gt;       - fill_by: Which column to use as the fill the dots. It must be a numeric column.</span></span>
<span><span class="co">#&gt;       If not provided, all dots will be filled with the same color at the middle of the palette.</span></span>
<span><span class="co">#&gt;       - fill_cutoff: A numeric value specifying the cutoff for the fill column.</span></span>
<span><span class="co">#&gt;       - fill_reverse: A logical value indicating whether to reverse the fill direction. Default is FALSE.</span></span>
<span><span class="co">#&gt;       By default, the fill direction is "up". If TRUE, the fill direction is "down".</span></span>
<span><span class="co">#&gt;       When the direction is "up", the values less than the cutoff will be filled with grey.</span></span>
<span><span class="co">#&gt;       When the direction is "down", the values greater than the cutoff will be filled with grey.</span></span>
<span><span class="co">#&gt;       - theme: A character string or a theme class (i.e. ggplot2::theme_classic) specifying the theme to use.</span></span>
<span><span class="co">#&gt;       Default is "theme_this".</span></span>
<span><span class="co">#&gt;       - theme_args: A list of arguments to pass to the theme function.</span></span>
<span><span class="co">#&gt;       - palette: A character string specifying the palette to use.</span></span>
<span><span class="co">#&gt;       A named list or vector can be used to specify the palettes for different split_by values.</span></span>
<span><span class="co">#&gt;       - palcolor: A character string specifying the color to use in the palette.</span></span>
<span><span class="co">#&gt;       A named list can be used to specify the colors for different split_by values.</span></span>
<span><span class="co">#&gt;       If some values are missing, the values from the palette will be used (palcolor will be NULL for those values).</span></span>
<span><span class="co">#&gt;       - alpha: A numeric value specifying the transparency of the plot.</span></span>
<span><span class="co">#&gt;       - facet_scales: Whether to scale the axes of facets. Default is "fixed"</span></span>
<span><span class="co">#&gt;       Other options are "free", "free_x", "free_y". See ggplot2::facet_wrap</span></span>
<span><span class="co">#&gt;       - facet_ncol: A numeric value specifying the number of columns in the facet.</span></span>
<span><span class="co">#&gt;       When facet_by is a single column and facet_wrap is used.</span></span>
<span><span class="co">#&gt;       - facet_nrow: A numeric value specifying the number of rows in the facet.</span></span>
<span><span class="co">#&gt;       When facet_by is a single column and facet_wrap is used.</span></span>
<span><span class="co">#&gt;       - facet_byrow: A logical value indicating whether to fill the plots by row. Default is TRUE.</span></span>
<span><span class="co">#&gt;       - seed: The random seed to use. Default is 8525.</span></span>
<span><span class="co">#&gt;       - aspect.ratio: A numeric value specifying the aspect ratio of the plot.</span></span>
<span><span class="co">#&gt;       - legend.position: A character string specifying the position of the legend.</span></span>
<span><span class="co">#&gt;       if waiver(), for single groups, the legend will be "none", otherwise "right".</span></span>
<span><span class="co">#&gt;       - legend.direction: A character string specifying the direction of the legend.</span></span>
<span><span class="co">#&gt;       - title: A character string specifying the title of the plot.</span></span>
<span><span class="co">#&gt;       A function can be used to generate the title based on the default title.</span></span>
<span><span class="co">#&gt;       This is useful when split_by is used and the title needs to be dynamic.</span></span>
<span><span class="co">#&gt;       - subtitle: A character string specifying the subtitle of the plot.</span></span>
<span><span class="co">#&gt;       - xlab: A character string specifying the x-axis label.</span></span>
<span><span class="co">#&gt;       - ylab: A character string specifying the y-axis label.</span></span>
<span><span class="co">#&gt;       - keep_na: A logical value or a character to replace the NA values in the data.</span></span>
<span><span class="co">#&gt;       It can also take a named list to specify different behavior for different columns.</span></span>
<span><span class="co">#&gt;       If TRUE or NA, NA values will be replaced with NA.</span></span>
<span><span class="co">#&gt;       If FALSE, NA values will be removed from the data before plotting.</span></span>
<span><span class="co">#&gt;       If a character string is provided, NA values will be replaced with the provided string.</span></span>
<span><span class="co">#&gt;       If a named vector/list is provided, the names should be the column names to apply the behavior to,</span></span>
<span><span class="co">#&gt;       and the values should be one of TRUE, FALSE, or a character string.</span></span>
<span><span class="co">#&gt;       Without a named vector/list, the behavior applies to categorical/character columns used on the plot,</span></span>
<span><span class="co">#&gt;       for example, the x, group_by, fill_by, etc.</span></span>
<span><span class="co">#&gt;       - keep_empty: One of FALSE, TRUE and "level". It can also take a named list to specify</span></span>
<span><span class="co">#&gt;       different behavior for different columns. Without a named list, the behavior applies to the</span></span>
<span><span class="co">#&gt;       categorical/character columns used on the plot, for example, the x, group_by, fill_by, etc.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       FALSE (default): Drop empty factor levels from the data before plotting.</span></span>
<span><span class="co">#&gt;       TRUE: Keep empty factor levels and show them as a separate category in the plot.</span></span>
<span><span class="co">#&gt;       "level": Keep empty factor levels, but do not show them in the plot.</span></span>
<span><span class="co">#&gt;       But they will be assigned colors from the palette to maintain consistency across multiple plots.</span></span>
<span><span class="co">#&gt;       Alias: levels</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - combine: Whether to combine the plots into one when facet is FALSE. Default is TRUE.</span></span>
<span><span class="co">#&gt;       - nrow: A numeric value specifying the number of rows in the facet.</span></span>
<span><span class="co">#&gt;       - ncol: A numeric value specifying the number of columns in the facet.</span></span>
<span><span class="co">#&gt;       - byrow: A logical value indicating whether to fill the plots by row.</span></span>
<span><span class="co">#&gt;       - axes: A string specifying how axes should be treated. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'keep' will retain all axes in individual plots.</span></span>
<span><span class="co">#&gt;        'collect' will remove duplicated axes when placed in the same run of rows or columns of the layout.</span></span>
<span><span class="co">#&gt;        'collect_x' and 'collect_y' will remove duplicated x-axes in the columns or duplicated y-axes in the rows respectively.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - axis_titles: A string specifying how axis titltes should be treated. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'keep' will retain all axis titles in individual plots.</span></span>
<span><span class="co">#&gt;        'collect' will remove duplicated titles in one direction and merge titles in the opposite direction.</span></span>
<span><span class="co">#&gt;        'collect_x' and 'collect_y' control this for x-axis titles and y-axis titles respectively.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - guides: A string specifying how guides should be treated in the layout. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'collect' will collect guides below to the given nesting level, removing duplicates.</span></span>
<span><span class="co">#&gt;        'keep' will stop collection at this level and let guides be placed alongside their plot.</span></span>
<span><span class="co">#&gt;        'auto' will allow guides to be collected if a upper level tries, but place them alongside the plot if not.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - design: Specification of the location of areas in the layout, passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE. When specified, nrow, ncol, and byrow are ignored.</span></span>
<span><span class="co">#&gt;       See patchwork::wrap_plots() for more details.</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt; - value</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;   A ggplot object or a list if combine is FALSE</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt; - examples</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;   set.seed(8525)</span></span>
<span><span class="co">#&gt;   data(cellphonedb_res)</span></span>
<span><span class="co">#&gt;   CCCPlot(data = cellphonedb_res, plot_type = "network", legend.position = "none",</span></span>
<span><span class="co">#&gt;     theme = "theme_blank", theme_args = list(add_coord = FALSE))</span></span>
<span><span class="co">#&gt;   CCCPlot(cellphonedb_res, plot_type = "chord")</span></span>
<span><span class="co">#&gt;   CCCPlot(cellphonedb_res, plot_type = "heatmap")</span></span>
<span><span class="co">#&gt;   CCCPlot(cellphonedb_res, plot_type = "dot",</span></span>
<span><span class="co">#&gt;     magnitude_agg = mean, magnitude_name = "Average Interaction Strength")</span></span>
<span><span class="co">#&gt;   CCCPlot(cellphonedb_res, plot_type = "sankey")</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;   cellphonedb_res_sub &lt;- cellphonedb_res[</span></span>
<span><span class="co">#&gt;     cellphonedb_res$source %in% c("Dendritic", "CD14+ Monocyte"),]</span></span>
<span><span class="co">#&gt;   CCCPlot(cellphonedb_res_sub, plot_type = "dot", method = "interaction")</span></span>
<span><span class="co">#&gt;   CCCPlot(cellphonedb_res_sub, plot_type = "network", method = "interaction",</span></span>
<span><span class="co">#&gt;     node_size_by = 1)</span></span>
<span><span class="co">#&gt;   CCCPlot(cellphonedb_res_sub, plot_type = "heatmap", method = "interaction",</span></span>
<span><span class="co">#&gt;     palette = "Reds")</span></span>
<span><span class="co">#&gt;   CCCPlot(cellphonedb_res_sub, plot_type = "box", method = "interaction")</span></span>
<span><span class="co">#&gt;   CCCPlot(cellphonedb_res_sub, plot_type = "violin", method = "interaction",</span></span>
<span><span class="co">#&gt;     add_box = TRUE)</span></span>
<span><span class="co">#&gt;   CCCPlot(cellphonedb_res_sub, plot_type = "ridge", method = "interaction")</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; You must code in the programming language 'R' to answer this prompt.</span></span>
<span><span class="co">#&gt; You can use functions from these packages: scplotter.</span></span>
<span><span class="co">#&gt; You may not install or load any additional packages.</span></span>
<span><span class="co">#&gt; These objects already exist in the R session:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Object_name, Type</span></span>
<span><span class="co">#&gt; cellphonedb_res, data.frame.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Do not define these objects in your R code.</span></span></code></pre>
<pre><code><span><span class="co">#&gt; <span style="font-weight: bold;">--- Receiving response from LLM provider: ---</span></span></span></code></pre>
<pre><code><span><span class="co">#&gt; ```r</span></span>
<span><span class="co">#&gt; CCCPlot(cellphonedb_res)</span></span>
<span><span class="co">#&gt; ```</span></span></code></pre>
<pre><code><span><span class="co">#&gt; Code ran:</span></span>
<span><span class="co">#&gt; CCCPlot(cellphonedb_res)</span></span></code></pre>
<p><img src="Visualizing_data_with_LLMs_files/figure-html/unnamed-chunk-10-1.png" class="r-plt" alt="" width="672"></p>
<p>To only debug a single conversation, you can set <code>verbose</code>
to <code>TRUE</code> in the <code>ask()</code> method.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chat</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/SCPlotterChat.html">SCPlotterChat</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span></span>
<span>    provider <span class="op">=</span> <span class="va">provider</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">chat</span><span class="op">$</span><span class="fu">ask</span><span class="op">(</span><span class="st">"Generate a cell-cell communication plot for the cellphonedb_res data."</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">--- Sending request to LLM provider (gpt-5-nano): ---</span></span></span>
<span><span class="co">#&gt; Objective: Select the most appropriate tool to handle the user's request while preserving conversational context. If the user refines or changes how the previous result should be visualized (e.g., asks for a different plot type), continue with the last plotting tool used unless they explicitly name a different tool.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Decision Process:</span></span>
<span><span class="co">#&gt; - If the user explicitly names a tool, output that tool.</span></span>
<span><span class="co">#&gt; - Else, if the request appears to refine the previous output (e.g., "do X instead", "make it a heatmap/dot/bar/etc", "change to ...", "add ... to", "same plot but ..."), select the last tool mentioned in the chat history.</span></span>
<span><span class="co">#&gt; - Else, analyze the current user request; if there is a clear, unambiguous match to a single tool, select that tool.</span></span>
<span><span class="co">#&gt; - Else, use the last mentioned tool from the chat history.</span></span>
<span><span class="co">#&gt; - If no tool is found, respond with "None".</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Response Format: Provide only the name of the selected tool, or "None" if no tool applies.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; User Request:</span></span>
<span><span class="co">#&gt; Generate a cell-cell communication plot for the cellphonedb_res data.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Available Tools:</span></span>
<span><span class="co">#&gt; - gt: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - CellDimPlot: Cell Dimension Reduction Plot  </span></span>
<span><span class="co">#&gt;    This function creates a dimension reduction plot for a Seurat object</span></span>
<span><span class="co">#&gt;    a Giotto object, a path to an .h5ad file or an opened    H5File    by    hdf5r    package.</span></span>
<span><span class="co">#&gt;    It allows for various customizations such as grouping by metadata,</span></span>
<span><span class="co">#&gt;    adding edges between cell neighbors, highlighting specific cells, and more.</span></span>
<span><span class="co">#&gt;    This function is a wrapper around    plotthis::DimPlot()   , which provides a</span></span>
<span><span class="co">#&gt;    flexible way to visualize cell clusters in reduced dimensions. This function</span></span>
<span><span class="co">#&gt;    extracts the necessary data from the Seurat or Giotto object and passes it to</span></span>
<span><span class="co">#&gt;    plotthis::DimPlot()   .</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalVolumePlot: ClonalVolumePlot  </span></span>
<span><span class="co">#&gt;    ClonalVolumePlot</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalRarefactionPlot: ClonalRarefactionPlot  </span></span>
<span><span class="co">#&gt;    Plot the rarefaction curves</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalStatPlot: ClonalStatPlot  </span></span>
<span><span class="co">#&gt;    Visualize the statistics of the clones.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalCompositionPlot: ClonalCompositionPlot  </span></span>
<span><span class="co">#&gt;    Plot the composition of the clones in different samples/groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - EnrichmentPlot: Enrichment Plot  </span></span>
<span><span class="co">#&gt;    This function generates various types of plots for enrichment (over-representation) analysis.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - eq: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - CellVelocityPlot: Cell Velocity Plot  </span></span>
<span><span class="co">#&gt;    This function creates a cell velocity plot for a Seurat object,</span></span>
<span><span class="co">#&gt;    a Giotto object, a path to an .h5ad file or an opened    H5File    by    hdf5r    package.</span></span>
<span><span class="co">#&gt;    It allows for various customizations such as grouping by metadata,</span></span>
<span><span class="co">#&gt;    adding edges between cell neighbors, highlighting specific cells, and more.</span></span>
<span><span class="co">#&gt;    This function is a wrapper around    plotthis::VelocityPlot()   , which provides a</span></span>
<span><span class="co">#&gt;    flexible way to visualize cell velocities in reduced dimensions. This function</span></span>
<span><span class="co">#&gt;    extracts the cell embeddings and velocity embeddings from the Seurat or Giotto object</span></span>
<span><span class="co">#&gt;    and passes them to    plotthis::VelocityPlot()   .</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - SpatFeaturePlot: Plot features for spatial data  </span></span>
<span><span class="co">#&gt;    The features can include  expression, dimension reduction components, metadata, etc</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - shared: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - CellStatPlot: Cell statistics plot  </span></span>
<span><span class="co">#&gt;    This function creates a plot to visualize the statistics of cells in a Seurat object, a Giotto object,</span></span>
<span><span class="co">#&gt;    a path to an .h5ad file or an opened    H5File    by    hdf5r    package.</span></span>
<span><span class="co">#&gt;    It can create various types of plots, including bar plots, circos plots, pie charts, pies (heatmap with cell_type = 'pie'), ring/donut plots, trend plots</span></span>
<span><span class="co">#&gt;    area plots, sankey/alluvial plots, heatmaps, radar plots, spider plots, violin plots, and box plots.</span></span>
<span><span class="co">#&gt;    The function allows for grouping, splitting, and faceting the data based on metadata columns.</span></span>
<span><span class="co">#&gt;    It also supports calculating fractions of cells based on specified groupings.#'</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ne: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - SpatDimPlot: Plot categories for spatial data  </span></span>
<span><span class="co">#&gt;    Plot categories for spatial data</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - GSEASummaryPlot: Objects exported from other packages  </span></span>
<span><span class="co">#&gt;    These objects are imported from other packages. Follow the links</span></span>
<span><span class="co">#&gt;    below to see their documentation.</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;         plotthis   GSEAPlot   ,    GSEASummaryPlot   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - CCCPlot: Cell-Cell Communication Plot  </span></span>
<span><span class="co">#&gt;    Plot the cell-cell communication.</span></span>
<span><span class="co">#&gt;    See also:</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;     The review:    https://www.sciencedirect.com/science/article/pii/S2452310021000081   </span></span>
<span><span class="co">#&gt;     The LIANA package:    https://liana-py.readthedocs.io/en/latest/notebooks/basic_usage.html#Tileplot   </span></span>
<span><span class="co">#&gt;     The CCPlotR package:    https://github.com/Sarah145/CCPlotR   </span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalResidencyPlot: ClonalResidencyPlot  </span></span>
<span><span class="co">#&gt;    Plot the residency of the clones in different samples.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClustreePlot: Clustree plot  </span></span>
<span><span class="co">#&gt;    This function generates a clustree plot from a data frame or a Seurat object.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalKmerPlot: ClonalKmerPlot  </span></span>
<span><span class="co">#&gt;    Explore the k-mer frequency of CDR3 sequences.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalPositionalPlot: ClonalPositionalPlot  </span></span>
<span><span class="co">#&gt;    Visualize the positional entropy, property or amino acid frequency of CDR3 sequences.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - uniq: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - top: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - le: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - MarkersPlot: Visualize Markers  </span></span>
<span><span class="co">#&gt;    Plot markers, typically identified by    Seurat::FindMarkers()    or    Seurat::FindAllMarkers()   .</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalOverlapPlot: ClonalOverlapPlot  </span></span>
<span><span class="co">#&gt;    Plot the overlap of the clones in different samples/groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ge: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalLengthPlot: ClonalLengthPlot  </span></span>
<span><span class="co">#&gt;    Plot the length distribution of the CDR3 sequences</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - and: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalAbundancePlot: ClonalAbundancePlot  </span></span>
<span><span class="co">#&gt;    Plot the count or density of the clones at different abundance levels.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - sel: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalDiversityPlot: ClonalDiversityPlot  </span></span>
<span><span class="co">#&gt;    Plot the clonal diversities of the samples/groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - FeatureStatPlot: Feature statistic plot  </span></span>
<span><span class="co">#&gt;    This function creates various types of feature statistic plots for a Seurat object, a Giotto object,</span></span>
<span><span class="co">#&gt;    a path to an .h5ad file or an opened    H5File    by    hdf5r    package.</span></span>
<span><span class="co">#&gt;    It allows for plotting features such as gene expression, scores, or other metadata across different groups or conditions.</span></span>
<span><span class="co">#&gt;    The function supports multiple plot types including violin, box, bar, ridge, dimension reduction, correlation, heatmap, and dot plots.</span></span>
<span><span class="co">#&gt;    It can also handle multiple features and supports faceting, splitting, and grouping by metadata columns.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ClonalGeneUsagePlot: ClonalGeneUsagePlot  </span></span>
<span><span class="co">#&gt;    ClonalGeneUsagePlot</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - lt: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - GSEAPlot: Objects exported from other packages  </span></span>
<span><span class="co">#&gt;    These objects are imported from other packages. Follow the links</span></span>
<span><span class="co">#&gt;    below to see their documentation.</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;         plotthis   GSEAPlot   ,    GSEASummaryPlot   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - or: Helper functions to select clones based on various criteria  </span></span>
<span><span class="co">#&gt;    These helper functions allow for the selection of clones based on various criteria such as size, group comparison, and existence in specific groups.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ListTools: List all available tools</span></span>
<span><span class="co">#&gt; List all available tools that can be used to handle the chat request.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; - ListData: List all available data objects</span></span>
<span><span class="co">#&gt; List all available data objects that can be used to handle the chat request.</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">--- Receiving response from LLM provider: ---</span></span></span>
<span><span class="co">#&gt; CCCPlot</span></span>
<span><span class="co">#&gt; Tool identified:  CCCPlot</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">--- Sending request to LLM provider (gpt-5-nano): ---</span></span></span>
<span><span class="co">#&gt; Objective: Identify the data object to be used.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Decision Process:</span></span>
<span><span class="co">#&gt; - If the user explicitly names a data object, output that data object name. If the name does not match exactly the available ones, use the one that is mostly related.</span></span>
<span><span class="co">#&gt; - Else, if the request appears to refine the previous output (e.g., "do X instead", "make it a heatmap/dot/bar/etc", "change to ...", "add ... to", "same plot but ..."), select the last dataset name mentioned in the chat history.</span></span>
<span><span class="co">#&gt; - Else, analyze the current user request; if there is a clear, unambiguous match to a dataset, select that data object.</span></span>
<span><span class="co">#&gt; - Else, use the last mentioned data object from the chat history.</span></span>
<span><span class="co">#&gt; - If no data object is found, respond with "None".</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Response Format: Provide only the name of the selected data object, or "None" if no tool applies. The response should be unquoted.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; User Request:</span></span>
<span><span class="co">#&gt; Generate a cell-cell communication plot for the cellphonedb_res data.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Available Data Objects:</span></span>
<span><span class="co">#&gt; - scplotter::cellphonedb_res: A toy example of CellPhoneDB output from LIANA</span></span>
<span><span class="co">#&gt; - scplotter::ifnb_sub: A subsetted version of 'ifnb' datasets</span></span>
<span><span class="co">#&gt; - scplotter::pancreas_sub: A subsetted version of mouse 'pancreas' datasets</span></span>
<span><span class="co">#&gt; - Seurat::cc.genes: Cell cycle genes</span></span>
<span><span class="co">#&gt; - Seurat::cc.genes.updated.2019: Cell cycle genes: 2019 update</span></span>
<span><span class="co">#&gt; - SeuratObject::pbmc_small: A small example version of the PBMC dataset</span></span>
<span><span class="co">#&gt; - scRepertoire::contig_list: A list of 8 single-cell T cell receptor sequences runs.</span></span>
<span><span class="co">#&gt; - scRepertoire::mini_contig_list: Processed subset of 'contig_list'</span></span>
<span><span class="co">#&gt; - scRepertoire::scRep_example: A Seurat object of 500 single T cells,</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">--- Receiving response from LLM provider: ---</span></span></span>
<span><span class="co">#&gt; scplotter::cellphonedb_res</span></span>
<span><span class="co">#&gt; Data object identified:  scplotter::cellphonedb_res</span></span>
<span><span class="co">#&gt; Warning in wrap$modify_fn(prompt_text, llm_provider): The 'skimr' package is</span></span>
<span><span class="co">#&gt; required to skim dataframes. Skim summary of dataframes currently not shown in</span></span>
<span><span class="co">#&gt; prompt</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">--- Sending request to LLM provider (gpt-5-nano): ---</span></span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Objective: Generate the R code to run the specified tool using the specified data object based on the user's request and the provided tool information.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Decision Process:</span></span>
<span><span class="co">#&gt; - Analyze the user's request to understand what needs to be done with the specified tool and data object.</span></span>
<span><span class="co">#&gt; - When use the data object name, do not quote it.</span></span>
<span><span class="co">#&gt; - Refer to the provided tool information to understand the tool's usage, arguments, and examples.</span></span>
<span><span class="co">#&gt; - Consider any specific parameters or options mentioned in the user's request that need to be included in the tool call.</span></span>
<span><span class="co">#&gt; - If the user's request is ambiguous or lacks necessary details, refer to the chat history for additional context that may help clarify the intended use of the tool and data object.</span></span>
<span><span class="co">#&gt; - Construct the R code that correctly calls the tool with the specified data object, ensuring that all necessary arguments are included and correctly formatted.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Response Format: Provide only the valid R code wrapped between "```r" and "```" to run the tool.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Tool to be used: CCCPlot</span></span>
<span><span class="co">#&gt; Data object to be used: cellphonedb_res</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; User Request:</span></span>
<span><span class="co">#&gt; Generate a cell-cell communication plot for the cellphonedb_res data.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Tool Information:</span></span>
<span><span class="co">#&gt; - title</span></span>
<span><span class="co">#&gt;   Cell-Cell Communication Plot</span></span>
<span><span class="co">#&gt; - description</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;    Plot the cell-cell communication.</span></span>
<span><span class="co">#&gt;    See also:</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;     The review:  https://www.sciencedirect.com/science/article/pii/S2452310021000081 </span></span>
<span><span class="co">#&gt;     The LIANA package:  https://liana-py.readthedocs.io/en/latest/notebooks/basic_usage.html#Tileplot </span></span>
<span><span class="co">#&gt;     The CCPlotR package:  https://github.com/Sarah145/CCPlotR </span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt; - usage</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;   CCCPlot(</span></span>
<span><span class="co">#&gt;     data,</span></span>
<span><span class="co">#&gt;     plot_type = c("dot", "network", "chord", "circos", "heatmap", "sankey", "alluvial",</span></span>
<span><span class="co">#&gt;       "box", "violin", "ridge"),</span></span>
<span><span class="co">#&gt;     method = c("aggregation", "interaction"),</span></span>
<span><span class="co">#&gt;     magnitude = waiver(),</span></span>
<span><span class="co">#&gt;     specificity = waiver(),</span></span>
<span><span class="co">#&gt;     magnitude_agg = length,</span></span>
<span><span class="co">#&gt;     magnitude_name = "No. of interactions",</span></span>
<span><span class="co">#&gt;     meta_specificity = "sumlog",</span></span>
<span><span class="co">#&gt;     split_by = NULL,</span></span>
<span><span class="co">#&gt;     x_text_angle = 90,</span></span>
<span><span class="co">#&gt;     link_curvature = 0.2,</span></span>
<span><span class="co">#&gt;     link_alpha = 0.6,</span></span>
<span><span class="co">#&gt;     facet_by = NULL,</span></span>
<span><span class="co">#&gt;     show_row_names = TRUE,</span></span>
<span><span class="co">#&gt;     show_column_names = TRUE,</span></span>
<span><span class="co">#&gt;     ...</span></span>
<span><span class="co">#&gt;   )</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt; - arguments</span></span>
<span><span class="co">#&gt;   - data: A data frame with the cell-cell communication data.</span></span>
<span><span class="co">#&gt;   A typical data frame should have the following columns:</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;    source The source cell type.</span></span>
<span><span class="co">#&gt;    target The target cell type.</span></span>
<span><span class="co">#&gt;    ligand The ligand gene.</span></span>
<span><span class="co">#&gt;    receptor The receptor gene.</span></span>
<span><span class="co">#&gt;    ligand_means The mean expression of the ligand gene per cell type.</span></span>
<span><span class="co">#&gt;    receptor_means The mean expression of the receptor gene per cell type.</span></span>
<span><span class="co">#&gt;    ligand_props The proportion of cells that express the entity.</span></span>
<span><span class="co">#&gt;    receptor_props The proportion of cells that express the entity.</span></span>
<span><span class="co">#&gt;    &lt;magnitude&gt; The magnitude of the communication.</span></span>
<span><span class="co">#&gt;    &lt;specificity&gt; The specificity of the communication.</span></span>
<span><span class="co">#&gt;   Depends on the plot_type, some columns are optional. But the source, target,</span></span>
<span><span class="co">#&gt;   ligand, receptor and &lt;magnitude&gt; are required.</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;   - plot_type: The type of plot to use. Default is "dot".</span></span>
<span><span class="co">#&gt;   Possible values are "network", "chord", "circos", "heatmap", "sankey", "alluvial", "dot",</span></span>
<span><span class="co">#&gt;   "box", "violin" and "ridge".</span></span>
<span><span class="co">#&gt;   For "box", "violin" and "ridge", the method should be "interaction".</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;    network: A network plot with the source and target cells as the nodes and the communication as the edges.</span></span>
<span><span class="co">#&gt;    chord: A chord plot with the source and target cells as the nodes and the communication as the chords.</span></span>
<span><span class="co">#&gt;    circos: Alias of "chord".</span></span>
<span><span class="co">#&gt;    heatmap: A heatmap plot with the source and target cells as the rows and columns.</span></span>
<span><span class="co">#&gt;    sankey: A sankey plot with the source and target cells as the nodes and the communication as the flows.</span></span>
<span><span class="co">#&gt;    alluvial: Alias of "sankey".</span></span>
<span><span class="co">#&gt;    dot: A dot plot with the source and target cells as the nodes and the communication as the dots.</span></span>
<span><span class="co">#&gt;    box: Box plots for source cell types. Each x is a target cell type and the values will be</span></span>
<span><span class="co">#&gt;   the interaction strengths of the ligand-receptor pairs.</span></span>
<span><span class="co">#&gt;    violin: Violin plots for source cell types. Each x is a target cell type and the values will be</span></span>
<span><span class="co">#&gt;   the interaction strengths of the ligand-receptor pairs.</span></span>
<span><span class="co">#&gt;    ridge: Ridge plots for source cell types. Each row is a target cell type and the values will be</span></span>
<span><span class="co">#&gt;   the interaction strengths of the ligand-receptor pairs.</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;   - method: The method to determine the plot entities.</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;    aggregation: Aggregate the ligand-receptor pairs interactions for each source-target pair.</span></span>
<span><span class="co">#&gt;   Only the source / target pairs will be plotted.</span></span>
<span><span class="co">#&gt;    interaction: Plot the ligand-receptor pairs interactions directly.</span></span>
<span><span class="co">#&gt;   The ligand-receptor pairs will also be plotted.</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;   - magnitude: The column name in the data to use as the magnitude of the communication.</span></span>
<span><span class="co">#&gt;   By default, the second last column will be used.</span></span>
<span><span class="co">#&gt;   See li.mt.show_methods() for the available methods in LIANA.</span></span>
<span><span class="co">#&gt;   or https://liana-py.readthedocs.io/en/latest/notebooks/basic_usage.html#Tileplot</span></span>
<span><span class="co">#&gt;   - specificity: The column name in the data to use as the specificity of the communication.</span></span>
<span><span class="co">#&gt;   By default, the last column will be used.</span></span>
<span><span class="co">#&gt;   If the method doesn't have a specificity, set it to NULL.</span></span>
<span><span class="co">#&gt;   - magnitude_agg: A function to aggregate the magnitude of the communication.</span></span>
<span><span class="co">#&gt;   Default is length.</span></span>
<span><span class="co">#&gt;   - magnitude_name: The name of the magnitude in the plot.</span></span>
<span><span class="co">#&gt;   Default is "No. of interactions".</span></span>
<span><span class="co">#&gt;   - meta_specificity: The method to calculate the specificity when there are multiple</span></span>
<span><span class="co">#&gt;   ligand-receptor pairs interactions. Default is "sumlog".</span></span>
<span><span class="co">#&gt;   It should be one of the methods in the metap package.</span></span>
<span><span class="co">#&gt;   Current available methods are:</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;    invchisq: Combine p values using the inverse chi squared method</span></span>
<span><span class="co">#&gt;    invt: Combine p values using the inverse t method</span></span>
<span><span class="co">#&gt;    logitp: Combine p values using the logit method</span></span>
<span><span class="co">#&gt;    meanp: Combine p values by the mean p method</span></span>
<span><span class="co">#&gt;    meanz: Combine p values using the mean z method</span></span>
<span><span class="co">#&gt;    sumlog: Combine p-values by the sum of logs (Fisher's) method</span></span>
<span><span class="co">#&gt;    sump: Combine p-values using the sum of p (Edgington's) method</span></span>
<span><span class="co">#&gt;    two2one: Convert two-sided p-values to one-sided</span></span>
<span><span class="co">#&gt;    votep: Combine p-values by the vote counting method</span></span>
<span><span class="co">#&gt;    wilkinsonp: Combine p-values using Wilkinson's method</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;   - split_by: A character vector of column names to split the plots. Default is NULL.</span></span>
<span><span class="co">#&gt;   - x_text_angle: The angle of the x-axis text. Default is 90.</span></span>
<span><span class="co">#&gt;   Only used when plot_type is "dot".</span></span>
<span><span class="co">#&gt;   - link_curvature: The curvature of the links. Default is 0.2.</span></span>
<span><span class="co">#&gt;   Only used when plot_type is "network".</span></span>
<span><span class="co">#&gt;   - link_alpha: The transparency of the links. Default is 0.6.</span></span>
<span><span class="co">#&gt;   Only used when plot_type is "network".</span></span>
<span><span class="co">#&gt;   - facet_by: A character vector of column names to facet the plots. Default is NULL.</span></span>
<span><span class="co">#&gt;   It should always be NULL.</span></span>
<span><span class="co">#&gt;   - show_row_names: Whether to show the row names in the heatmap. Default is TRUE.</span></span>
<span><span class="co">#&gt;   Only used when plot_type is "heatmap".</span></span>
<span><span class="co">#&gt;   - show_column_names: Whether to show the column names in the heatmap. Default is TRUE.</span></span>
<span><span class="co">#&gt;   Only used when plot_type is "heatmap".</span></span>
<span><span class="co">#&gt;   - ...: Other arguments passed to the specific plot function.</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;    For Network, see plotthis::Network().</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;       [...] can be:</span></span>
<span><span class="co">#&gt;       - links: A data frame containing the links between nodes.</span></span>
<span><span class="co">#&gt;       - nodes: A data frame containing the nodes.</span></span>
<span><span class="co">#&gt;       This is optional. The names of the nodes are extracted from the links data frame.</span></span>
<span><span class="co">#&gt;       If "@nodes" is provided, the nodes data frame will be extracted from the attribute nodes of the links data frame.</span></span>
<span><span class="co">#&gt;       - split_by_sep: The separator for multiple split_by columns. See split_by</span></span>
<span><span class="co">#&gt;       - split_nodes: A logical value specifying whether to split the nodes data.</span></span>
<span><span class="co">#&gt;       If TRUE, the nodes data will also be split by the split_by column.</span></span>
<span><span class="co">#&gt;       - from: A character string specifying the column name of the links data frame for the source nodes.</span></span>
<span><span class="co">#&gt;       Default is the first column of the links data frame.</span></span>
<span><span class="co">#&gt;       - from_sep: A character string to concatenate the columns in from, if multiple columns are provided.</span></span>
<span><span class="co">#&gt;       - to: A character string specifying the column name of the links data frame for the target nodes.</span></span>
<span><span class="co">#&gt;       Default is the second column of the links data frame.</span></span>
<span><span class="co">#&gt;       - to_sep: A character string to concatenate the columns in to, if multiple columns are provided.</span></span>
<span><span class="co">#&gt;       - node_by: A character string specifying the column name of the nodes data frame for the node names.</span></span>
<span><span class="co">#&gt;       Default is the first column of the nodes data frame.</span></span>
<span><span class="co">#&gt;       - node_by_sep: A character string to concatenate the columns in node_by, if multiple columns are provided.</span></span>
<span><span class="co">#&gt;       - link_weight_by: A numeric value or a character string specifying the column name of the links data frame for the link weight.</span></span>
<span><span class="co">#&gt;       If a numeric value is provided, all links will have the same weight.</span></span>
<span><span class="co">#&gt;       This determines the width of the links.</span></span>
<span><span class="co">#&gt;       - link_weight_name: A character string specifying the name of the link weight in the legend.</span></span>
<span><span class="co">#&gt;       - link_type_by: A character string specifying the type of the links.</span></span>
<span><span class="co">#&gt;       This can be "solid", "dashed", "dotted", or a column name from the links data frame.</span></span>
<span><span class="co">#&gt;       It has higher priority when it is a column name.</span></span>
<span><span class="co">#&gt;       - link_type_name: A character string specifying the name of the link type in the legend.</span></span>
<span><span class="co">#&gt;       - node_size_by: A numeric value or a character string specifying the column name of the nodes data frame for the node size.</span></span>
<span><span class="co">#&gt;       If a numeric value is provided, all nodes will have the same size.</span></span>
<span><span class="co">#&gt;       - node_size_name: A character string specifying the name of the node size in the legend.</span></span>
<span><span class="co">#&gt;       - node_color_by: A character string specifying the color of the nodes.</span></span>
<span><span class="co">#&gt;       This can be a color name, a hex code, or a column name from the nodes data frame.</span></span>
<span><span class="co">#&gt;       It has higher priority when it is a column name.</span></span>
<span><span class="co">#&gt;       - node_color_name: A character string specifying the name of the node color in the legend.</span></span>
<span><span class="co">#&gt;       - node_shape_by: A numeric value or a character string specifying the column name of the nodes data frame for the node shape.</span></span>
<span><span class="co">#&gt;       If a numeric value is provided, all nodes will have the same shape.</span></span>
<span><span class="co">#&gt;       - node_shape_name: A character string specifying the name of the node shape in the legend.</span></span>
<span><span class="co">#&gt;       - node_fill_by: A character string specifying the fill color of the nodes.</span></span>
<span><span class="co">#&gt;       This can be a color name, a hex code, or a column name from the nodes data frame.</span></span>
<span><span class="co">#&gt;       It has higher priority when it is a column name.</span></span>
<span><span class="co">#&gt;       - node_fill_name: A character string specifying the name of the node fill in the legend.</span></span>
<span><span class="co">#&gt;       - node_alpha: A numeric value specifying the transparency of the nodes.</span></span>
<span><span class="co">#&gt;       It only works when the nodes are filled.</span></span>
<span><span class="co">#&gt;       - node_stroke: A numeric value specifying the stroke of the nodes.</span></span>
<span><span class="co">#&gt;       - cluster_scale: A character string specifying how to scale the clusters.</span></span>
<span><span class="co">#&gt;       It can be "fill", "color", or "shape".</span></span>
<span><span class="co">#&gt;       - node_size_range: A numeric vector specifying the range of the node size.</span></span>
<span><span class="co">#&gt;       - link_weight_range: A numeric vector specifying the range of the link weight.</span></span>
<span><span class="co">#&gt;       - link_arrow_offset: A numeric value specifying the offset of the link arrows.</span></span>
<span><span class="co">#&gt;       So that they won't overlap with the nodes.</span></span>
<span><span class="co">#&gt;       - link_color_by: A character string specifying the colors of the link. It can be:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        "from" means the color of the link is determined by the source node.</span></span>
<span><span class="co">#&gt;        "to" means the color of the link is determined by the target node.</span></span>
<span><span class="co">#&gt;        Otherwise, the color of the link is determined by the column name from the links data frame.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - link_color_name: A character string specifying the name of the link color in the legend.</span></span>
<span><span class="co">#&gt;       Only used when link_color_by is a column name.</span></span>
<span><span class="co">#&gt;       - palette: A character string specifying the palette to use.</span></span>
<span><span class="co">#&gt;       A named list or vector can be used to specify the palettes for different split_by values.</span></span>
<span><span class="co">#&gt;       - palcolor: A character string specifying the color to use in the palette.</span></span>
<span><span class="co">#&gt;       A named list can be used to specify the colors for different split_by values.</span></span>
<span><span class="co">#&gt;       If some values are missing, the values from the palette will be used (palcolor will be NULL for those values).</span></span>
<span><span class="co">#&gt;       - link_palette: A character string specifying the palette of the links.</span></span>
<span><span class="co">#&gt;       When link_color_by is "from" or "to", the palette of the links defaults to the palette of the nodes.</span></span>
<span><span class="co">#&gt;       - link_palcolor: A character vector specifying the colors of the link palette.</span></span>
<span><span class="co">#&gt;       When link_color_by is "from" or "to", the colors of the link palette defaults to the colors of the node palette.</span></span>
<span><span class="co">#&gt;       - directed: A logical value specifying whether the graph is directed.</span></span>
<span><span class="co">#&gt;       - layout: A character string specifying the layout of the graph.</span></span>
<span><span class="co">#&gt;       It can be "circle", "tree", "grid", or a layout function from igraph.</span></span>
<span><span class="co">#&gt;       - cluster: A character string specifying the clustering method.</span></span>
<span><span class="co">#&gt;       It can be "none", "fast_greedy", "walktrap", "edge_betweenness", "infomap", or a clustering function from igraph.</span></span>
<span><span class="co">#&gt;       - add_mark: A logical value specifying whether to add mark for the clusters to the plot.</span></span>
<span><span class="co">#&gt;       - mark_expand: A unit value specifying the expansion of the mark.</span></span>
<span><span class="co">#&gt;       - mark_type: A character string specifying the type of the mark.</span></span>
<span><span class="co">#&gt;       It can be "hull", "ellipse", "rect", "circle", or a mark function from ggforce.</span></span>
<span><span class="co">#&gt;       - mark_alpha: A numeric value specifying the transparency of the mark.</span></span>
<span><span class="co">#&gt;       - mark_linetype: A numeric value specifying the line type of the mark.</span></span>
<span><span class="co">#&gt;       - add_label: A logical value specifying whether to add label to the nodes to the plot.</span></span>
<span><span class="co">#&gt;       - label_size: A numeric value specifying the size of the label.</span></span>
<span><span class="co">#&gt;       - label_fg: A character string specifying the foreground color of the label.</span></span>
<span><span class="co">#&gt;       - label_bg: A character string specifying the background color of the label.</span></span>
<span><span class="co">#&gt;       - label_bg_r: A numeric value specifying the background ratio of the label.</span></span>
<span><span class="co">#&gt;       - arrow: An arrow object for the links.</span></span>
<span><span class="co">#&gt;       - title: A character string specifying the title of the plot.</span></span>
<span><span class="co">#&gt;       A function can be used to generate the title based on the default title.</span></span>
<span><span class="co">#&gt;       This is useful when split_by is used and the title needs to be dynamic.</span></span>
<span><span class="co">#&gt;       - subtitle: A character string specifying the subtitle of the plot.</span></span>
<span><span class="co">#&gt;       - xlab: A character string specifying the x-axis label.</span></span>
<span><span class="co">#&gt;       - ylab: A character string specifying the y-axis label.</span></span>
<span><span class="co">#&gt;       - aspect.ratio: A numeric value specifying the aspect ratio of the plot.</span></span>
<span><span class="co">#&gt;       - theme: A character string or a theme class (i.e. ggplot2::theme_classic) specifying the theme to use.</span></span>
<span><span class="co">#&gt;       Default is "theme_this".</span></span>
<span><span class="co">#&gt;       - theme_args: A list of arguments to pass to the theme function.</span></span>
<span><span class="co">#&gt;       - legend.position: A character string specifying the position of the legend.</span></span>
<span><span class="co">#&gt;       if waiver(), for single groups, the legend will be "none", otherwise "right".</span></span>
<span><span class="co">#&gt;       - legend.direction: A character string specifying the direction of the legend.</span></span>
<span><span class="co">#&gt;       - seed: The random seed to use. Default is 8525.</span></span>
<span><span class="co">#&gt;       - combine: Whether to combine the plots into one when facet is FALSE. Default is TRUE.</span></span>
<span><span class="co">#&gt;       - nrow: A numeric value specifying the number of rows in the facet.</span></span>
<span><span class="co">#&gt;       - ncol: A numeric value specifying the number of columns in the facet.</span></span>
<span><span class="co">#&gt;       - byrow: A logical value indicating whether to fill the plots by row.</span></span>
<span><span class="co">#&gt;       - axes: A string specifying how axes should be treated. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'keep' will retain all axes in individual plots.</span></span>
<span><span class="co">#&gt;        'collect' will remove duplicated axes when placed in the same run of rows or columns of the layout.</span></span>
<span><span class="co">#&gt;        'collect_x' and 'collect_y' will remove duplicated x-axes in the columns or duplicated y-axes in the rows respectively.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - axis_titles: A string specifying how axis titltes should be treated. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'keep' will retain all axis titles in individual plots.</span></span>
<span><span class="co">#&gt;        'collect' will remove duplicated titles in one direction and merge titles in the opposite direction.</span></span>
<span><span class="co">#&gt;        'collect_x' and 'collect_y' control this for x-axis titles and y-axis titles respectively.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - guides: A string specifying how guides should be treated in the layout. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'collect' will collect guides below to the given nesting level, removing duplicates.</span></span>
<span><span class="co">#&gt;        'keep' will stop collection at this level and let guides be placed alongside their plot.</span></span>
<span><span class="co">#&gt;        'auto' will allow guides to be collected if a upper level tries, but place them alongside the plot if not.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - design: Specification of the location of areas in the layout, passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE. When specified, nrow, ncol, and byrow are ignored.</span></span>
<span><span class="co">#&gt;       See patchwork::wrap_plots() for more details.</span></span>
<span><span class="co">#&gt;   For ChordPlot, see plotthis::ChordPlot().</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;       [...] can be:</span></span>
<span><span class="co">#&gt;       - y: A character string specifying the column name of the data frame to plot for the y-axis.</span></span>
<span><span class="co">#&gt;       - from: A character string of the column name to plot for the source.</span></span>
<span><span class="co">#&gt;       A character/factor column is expected.</span></span>
<span><span class="co">#&gt;       - from_sep: A character string to concatenate the columns in from, if multiple columns are provided.</span></span>
<span><span class="co">#&gt;       - to: A character string of the column name to plot for the target.</span></span>
<span><span class="co">#&gt;       A character/factor column is expected.</span></span>
<span><span class="co">#&gt;       - to_sep: A character string to concatenate the columns in to, if multiple columns are provided.</span></span>
<span><span class="co">#&gt;       - split_by_sep: The separator for multiple split_by columns. See split_by</span></span>
<span><span class="co">#&gt;       - flip: A logical value to flip the source and target.</span></span>
<span><span class="co">#&gt;       - links_color: A character string to specify the color of the links.</span></span>
<span><span class="co">#&gt;       Either "from" or "to".</span></span>
<span><span class="co">#&gt;       - theme: A character string or a theme class (i.e. ggplot2::theme_classic) specifying the theme to use.</span></span>
<span><span class="co">#&gt;       Default is "theme_this".</span></span>
<span><span class="co">#&gt;       - theme_args: A list of arguments to pass to the theme function.</span></span>
<span><span class="co">#&gt;       - palette: A character string specifying the palette to use.</span></span>
<span><span class="co">#&gt;       A named list or vector can be used to specify the palettes for different split_by values.</span></span>
<span><span class="co">#&gt;       - palcolor: A character string specifying the color to use in the palette.</span></span>
<span><span class="co">#&gt;       A named list can be used to specify the colors for different split_by values.</span></span>
<span><span class="co">#&gt;       If some values are missing, the values from the palette will be used (palcolor will be NULL for those values).</span></span>
<span><span class="co">#&gt;       - alpha: A numeric value specifying the transparency of the plot.</span></span>
<span><span class="co">#&gt;       - labels_rot: A logical value to rotate the labels by 90 degrees.</span></span>
<span><span class="co">#&gt;       - title: A character string specifying the title of the plot.</span></span>
<span><span class="co">#&gt;       A function can be used to generate the title based on the default title.</span></span>
<span><span class="co">#&gt;       This is useful when split_by is used and the title needs to be dynamic.</span></span>
<span><span class="co">#&gt;       - subtitle: A character string specifying the subtitle of the plot.</span></span>
<span><span class="co">#&gt;       - seed: The random seed to use. Default is 8525.</span></span>
<span><span class="co">#&gt;       - keep_na: A logical value or a character to replace the NA values in the data.</span></span>
<span><span class="co">#&gt;       It can also take a named list to specify different behavior for different columns.</span></span>
<span><span class="co">#&gt;       If TRUE or NA, NA values will be replaced with NA.</span></span>
<span><span class="co">#&gt;       If FALSE, NA values will be removed from the data before plotting.</span></span>
<span><span class="co">#&gt;       If a character string is provided, NA values will be replaced with the provided string.</span></span>
<span><span class="co">#&gt;       If a named vector/list is provided, the names should be the column names to apply the behavior to,</span></span>
<span><span class="co">#&gt;       and the values should be one of TRUE, FALSE, or a character string.</span></span>
<span><span class="co">#&gt;       Without a named vector/list, the behavior applies to categorical/character columns used on the plot,</span></span>
<span><span class="co">#&gt;       for example, the x, group_by, fill_by, etc.</span></span>
<span><span class="co">#&gt;       - keep_empty: One of FALSE, TRUE and "level". It can also take a named list to specify</span></span>
<span><span class="co">#&gt;       different behavior for different columns. Without a named list, the behavior applies to the</span></span>
<span><span class="co">#&gt;       categorical/character columns used on the plot, for example, the x, group_by, fill_by, etc.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       FALSE (default): Drop empty factor levels from the data before plotting.</span></span>
<span><span class="co">#&gt;       TRUE: Keep empty factor levels and show them as a separate category in the plot.</span></span>
<span><span class="co">#&gt;       "level": Keep empty factor levels, but do not show them in the plot.</span></span>
<span><span class="co">#&gt;       But they will be assigned colors from the palette to maintain consistency across multiple plots.</span></span>
<span><span class="co">#&gt;       Alias: levels</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - combine: Whether to combine the plots into one when facet is FALSE. Default is TRUE.</span></span>
<span><span class="co">#&gt;       - nrow: A numeric value specifying the number of rows in the facet.</span></span>
<span><span class="co">#&gt;       - ncol: A numeric value specifying the number of columns in the facet.</span></span>
<span><span class="co">#&gt;       - byrow: A logical value indicating whether to fill the plots by row.</span></span>
<span><span class="co">#&gt;       - axes: A string specifying how axes should be treated. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'keep' will retain all axes in individual plots.</span></span>
<span><span class="co">#&gt;        'collect' will remove duplicated axes when placed in the same run of rows or columns of the layout.</span></span>
<span><span class="co">#&gt;        'collect_x' and 'collect_y' will remove duplicated x-axes in the columns or duplicated y-axes in the rows respectively.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - axis_titles: A string specifying how axis titltes should be treated. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'keep' will retain all axis titles in individual plots.</span></span>
<span><span class="co">#&gt;        'collect' will remove duplicated titles in one direction and merge titles in the opposite direction.</span></span>
<span><span class="co">#&gt;        'collect_x' and 'collect_y' control this for x-axis titles and y-axis titles respectively.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - guides: A string specifying how guides should be treated in the layout. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'collect' will collect guides below to the given nesting level, removing duplicates.</span></span>
<span><span class="co">#&gt;        'keep' will stop collection at this level and let guides be placed alongside their plot.</span></span>
<span><span class="co">#&gt;        'auto' will allow guides to be collected if a upper level tries, but place them alongside the plot if not.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - design: Specification of the location of areas in the layout, passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE. When specified, nrow, ncol, and byrow are ignored.</span></span>
<span><span class="co">#&gt;       See patchwork::wrap_plots() for more details.</span></span>
<span><span class="co">#&gt;   For Heatmap, see plotthis::Heatmap().</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;       [...] can be:</span></span>
<span><span class="co">#&gt;       - values_by: A character of column name in data that contains the values to be plotted.</span></span>
<span><span class="co">#&gt;       This is required when in_form is "long". For other formats, the values are pivoted into a column named by values_by.</span></span>
<span><span class="co">#&gt;       - values_fill: A value to fill in the missing values in the heatmap.</span></span>
<span><span class="co">#&gt;       When there is missing value in the data, the cluster_rows and cluster_columns will fail.</span></span>
<span><span class="co">#&gt;       - name: A character string to name the heatmap (will be used to rename values_by).</span></span>
<span><span class="co">#&gt;       - in_form: The format of the data. Can be one of "matrix", "long", "wide-rows", "wide-columns", or "auto".</span></span>
<span><span class="co">#&gt;       Defaults to "auto".</span></span>
<span><span class="co">#&gt;       - split_by_sep: A character string to concat multiple columns in split_by.</span></span>
<span><span class="co">#&gt;       - rows_by: A vector of column names in data that contains the row information.</span></span>
<span><span class="co">#&gt;       This is used to create the rows of the heatmap.</span></span>
<span><span class="co">#&gt;       When in_form is "long" or "wide-columns", this is requied, and multiple columns can be specified,</span></span>
<span><span class="co">#&gt;       which will be concatenated by rows_by_sep into a single column.</span></span>
<span><span class="co">#&gt;       - rows_by_sep: A character string to concat multiple columns in rows_by.</span></span>
<span><span class="co">#&gt;       - rows_split_by: A character of column name in data that contains the split information for rows.</span></span>
<span><span class="co">#&gt;       - rows_split_by_sep: A character string to concat multiple columns in rows_split_by.</span></span>
<span><span class="co">#&gt;       - columns_by: A vector of column names in data that contains the column information.</span></span>
<span><span class="co">#&gt;       This is used to create the columns of the heatmap.</span></span>
<span><span class="co">#&gt;       When in_form is "long" or "wide-rows", this is required, and multiple columns can be specified,</span></span>
<span><span class="co">#&gt;       which will be concatenated by columns_by_sep into a single column.</span></span>
<span><span class="co">#&gt;       - columns_by_sep: A character string to concat multiple columns in columns_by.</span></span>
<span><span class="co">#&gt;       - columns_split_by: A character of column name in data that contains the split information for columns.</span></span>
<span><span class="co">#&gt;       - columns_split_by_sep: A character string to concat multiple columns in columns_split_by.</span></span>
<span><span class="co">#&gt;       - rows_data: A data frame containing additional data for rows, which can be used to add annotations to the heatmap.</span></span>
<span><span class="co">#&gt;       It will be joined to the main data by rows_by and split_by if split_by exists in rows_data.</span></span>
<span><span class="co">#&gt;       This is useful for adding additional information to the rows of the heatmap.</span></span>
<span><span class="co">#&gt;       - columns_data: A data frame containing additional data for columns, which can be used to add annotations to the heatmap.</span></span>
<span><span class="co">#&gt;       It will be joined to the main data by columns_by and split_by if split_by exists in columns_data.</span></span>
<span><span class="co">#&gt;       This is useful for adding additional information to the columns of the heatmap.</span></span>
<span><span class="co">#&gt;       - keep_na: Whether we should keep NA groups in rows, columns and split_by variables. Default is FALSE.</span></span>
<span><span class="co">#&gt;       FALSE to remove NA groups; TRUE to keep NA groups.</span></span>
<span><span class="co">#&gt;       A vector of column names can also be provided to specify which columns to keep NA groups.</span></span>
<span><span class="co">#&gt;       Note that the record will be removed if any of the grouping columns has NA and is not specified to keep NA.</span></span>
<span><span class="co">#&gt;       - keep_empty: One of FALSE, TRUE and "level". It can also take a named list to specify</span></span>
<span><span class="co">#&gt;       different behavior for different columns. Without a named list, the behavior applies to the</span></span>
<span><span class="co">#&gt;       categorical/character columns used on the plot, for example, the x, group_by, fill_by, etc.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       FALSE (default): Drop empty factor levels from the data before plotting.</span></span>
<span><span class="co">#&gt;       TRUE: Keep empty factor levels and show them as a separate category in the plot.</span></span>
<span><span class="co">#&gt;       "level": Keep empty factor levels, but do not show them in the plot.</span></span>
<span><span class="co">#&gt;       But they will be assigned colors from the palette to maintain consistency across multiple plots.</span></span>
<span><span class="co">#&gt;       Alias: levels</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - rows_orderby: A expression (in character) to specify how to order rows. It will be evaluated in the context of the data frame used for rows (after grouping by rows_split_by and rows_by). The expression should return a vector of the same length as the number of rows in the data frame. The default is NULL, which means no specific ordering.</span></span>
<span><span class="co">#&gt;       Can't be used with cluster_rows = TRUE.</span></span>
<span><span class="co">#&gt;       This is applied before renaming rows_by to rows_name.</span></span>
<span><span class="co">#&gt;       - columns_orderby: A expression (in character) to specify how to order columns. It will be evaluated in the context of the data frame used for columns (after grouping by columns</span></span>
<span><span class="co">#&gt;       split_by and columns_by). The expression should return a vector of the same length as the number of rows in the data frame. The default is NULL, which means no specific ordering.</span></span>
<span><span class="co">#&gt;       Can't be used with cluster_columns = TRUE.</span></span>
<span><span class="co">#&gt;       This is applied before renaming columns_by to columns_name.</span></span>
<span><span class="co">#&gt;       - columns_name: A character string to rename the column created by columns_by, which will be reflected in the name of the annotation or legend.</span></span>
<span><span class="co">#&gt;       - columns_split_name: A character string to rename the column created by columns_split_by, which will be reflected in the name of the annotation or legend.</span></span>
<span><span class="co">#&gt;       - rows_name: A character string to rename the column created by rows_by, which will be reflected in the name of the annotation or legend.</span></span>
<span><span class="co">#&gt;       - rows_split_name: A character string to rename the column created by rows_split_by, which will be reflected in the name of the annotation or legend.</span></span>
<span><span class="co">#&gt;       - palette: A character string specifying the palette of the heatmap cells.</span></span>
<span><span class="co">#&gt;       - palcolor: A character vector of colors to override the palette of the heatmap cells.</span></span>
<span><span class="co">#&gt;       - rows_palette: A character string specifying the palette of the row group annotation.</span></span>
<span><span class="co">#&gt;       The default is "Paired".</span></span>
<span><span class="co">#&gt;       - rows_palcolor: A character vector of colors to override the palette of the row group annotation.</span></span>
<span><span class="co">#&gt;       - rows_split_palette: A character string specifying the palette of the row split annotation.</span></span>
<span><span class="co">#&gt;       The default is "simspec".</span></span>
<span><span class="co">#&gt;       - rows_split_palcolor: A character vector of colors to override the palette of the row split annotation.</span></span>
<span><span class="co">#&gt;       - columns_palette: A character string specifying the palette of the column group annotation.</span></span>
<span><span class="co">#&gt;       The default is "Paired".</span></span>
<span><span class="co">#&gt;       - columns_palcolor: A character vector of colors to override the palette of the column group annotation.</span></span>
<span><span class="co">#&gt;       - columns_split_palette: A character string specifying the palette of the column split annotation.</span></span>
<span><span class="co">#&gt;       The default is "simspec".</span></span>
<span><span class="co">#&gt;       - columns_split_palcolor: A character vector of colors to override the palette of the column split annotation.</span></span>
<span><span class="co">#&gt;       - pie_size_name: A character string specifying the name of the legend for the pie size.</span></span>
<span><span class="co">#&gt;       - pie_size: A numeric value or a function specifying the size of the pie chart.</span></span>
<span><span class="co">#&gt;       If it is a function, the function should take count as the argument and return the size.</span></span>
<span><span class="co">#&gt;       - pie_values: A function or character that can be converted to a function by match.arg()</span></span>
<span><span class="co">#&gt;       to calculate the values for the pie chart. Default is "length".</span></span>
<span><span class="co">#&gt;       The function should take a vector of values as the argument and return a single value, for each</span></span>
<span><span class="co">#&gt;       group in pie_group_by.</span></span>
<span><span class="co">#&gt;       - pie_name: A character string to rename the column created by pie_group_by, which will be reflected in the name of the annotation or legend.</span></span>
<span><span class="co">#&gt;       - pie_group_by: A character of column name in data that contains the group information for pie charts.</span></span>
<span><span class="co">#&gt;       This is used to create pie charts in the heatmap when cell_type is "pie".</span></span>
<span><span class="co">#&gt;       - pie_group_by_sep: A character string to concat multiple columns in pie_group_by.</span></span>
<span><span class="co">#&gt;       - pie_palette: A character string specifying the palette of the pie chart.</span></span>
<span><span class="co">#&gt;       - pie_palcolor: A character vector of colors to override the palette of the pie chart.</span></span>
<span><span class="co">#&gt;       - bars_sample: An integer specifying the number of samples to draw the bars.</span></span>
<span><span class="co">#&gt;       - label: A function to calculate the labels for the heatmap cells.</span></span>
<span><span class="co">#&gt;       It can take either 1, 3, or 5 arguments. The first argument is the aggregated values.</span></span>
<span><span class="co">#&gt;       If it takes 3 arguments, the second and third arguments are the row and column indices.</span></span>
<span><span class="co">#&gt;       If it takes 5 arguments, the second and third arguments are the row and column indices,</span></span>
<span><span class="co">#&gt;       the fourth and fifth arguments are the row and column names.</span></span>
<span><span class="co">#&gt;       The function should return a character vector of the same length as the aggregated values.</span></span>
<span><span class="co">#&gt;       If the function returns NA, no label will be shown for that cell.</span></span>
<span><span class="co">#&gt;       For the indices, if you have the same dimension of data (same order of rows and columns) as the heatmap, you need to use ComplexHeatmap::pindex() to get the correct values.</span></span>
<span><span class="co">#&gt;       - label_size: A numeric value specifying the size of the labels when cell_type = "label".</span></span>
<span><span class="co">#&gt;       - violin_fill: A character vector of colors to override the fill color of the violin plot.</span></span>
<span><span class="co">#&gt;       If NULL, the fill color will be the same as the annotion.</span></span>
<span><span class="co">#&gt;       - boxplot_fill: A character vector of colors to override the fill color of the boxplot.</span></span>
<span><span class="co">#&gt;       If NULL, the fill color will be the same as the annotion.</span></span>
<span><span class="co">#&gt;       - dot_size: A numeric value specifying the size of the dot or a function to calculate the size</span></span>
<span><span class="co">#&gt;       from the values in the cell or a function to calculate the size from the values in the cell.</span></span>
<span><span class="co">#&gt;       - dot_size_name: A character string specifying the name of the legend for the dot size.</span></span>
<span><span class="co">#&gt;       If NULL, the dot size legend will not be shown.</span></span>
<span><span class="co">#&gt;       - legend_items: A numeric vector with names to specifiy the items in the main legend.</span></span>
<span><span class="co">#&gt;       The names will be working as the labels of the legend items.</span></span>
<span><span class="co">#&gt;       - legend_discrete: A logical value indicating whether the main legend is discrete.</span></span>
<span><span class="co">#&gt;       - legend.position: A character string specifying the position of the legend.</span></span>
<span><span class="co">#&gt;       if waiver(), for single groups, the legend will be "none", otherwise "right".</span></span>
<span><span class="co">#&gt;       - legend.direction: A character string specifying the direction of the legend.</span></span>
<span><span class="co">#&gt;       - lower_quantile, upper_quantile, lower_cutoff, upper_cutoff: Vector of minimum and maximum cutoff values or quantile values for each feature.</span></span>
<span><span class="co">#&gt;       It's applied to aggregated values when aggregated values are used (e.g. plot_type tile, label, etc).</span></span>
<span><span class="co">#&gt;       It's applied to raw values when raw values are used (e.g. plot_type bars, etc).</span></span>
<span><span class="co">#&gt;       - add_bg: A logical value indicating whether to add a background to the heatmap.</span></span>
<span><span class="co">#&gt;       Does not work with cell_type = "bars" or cell_type = "tile".</span></span>
<span><span class="co">#&gt;       - bg_alpha: A numeric value between 0 and 1 specifying the transparency of the background.</span></span>
<span><span class="co">#&gt;       - add_reticle: A logical value indicating whether to add a reticle to the heatmap.</span></span>
<span><span class="co">#&gt;       - reticle_color: A character string specifying the color of the reticle.</span></span>
<span><span class="co">#&gt;       - column_name_annotation: A logical value indicating whether to add the column annotation for the column names.</span></span>
<span><span class="co">#&gt;       which is a simple annotaion indicating the column names.</span></span>
<span><span class="co">#&gt;       - column_name_legend: A logical value indicating whether to show the legend of the column name annotation.</span></span>
<span><span class="co">#&gt;       - row_name_annotation: A logical value indicating whether to add the row annotation for the row names.</span></span>
<span><span class="co">#&gt;       which is a simple annotaion indicating the row names.</span></span>
<span><span class="co">#&gt;       - row_name_legend: A logical value indicating whether to show the legend of the row name annotation.</span></span>
<span><span class="co">#&gt;       - cluster_columns: A logical value indicating whether to cluster the columns.</span></span>
<span><span class="co">#&gt;       If TRUE and columns_split_by is provided, the clustering will only be applied to the columns within the same split.</span></span>
<span><span class="co">#&gt;       - cluster_rows: A logical value indicating whether to cluster the rows.</span></span>
<span><span class="co">#&gt;       If TRUE and rows_split_by is provided, the clustering will only be applied to the rows within the same split.</span></span>
<span><span class="co">#&gt;       - border: A logical value indicating whether to draw the border of the heatmap.</span></span>
<span><span class="co">#&gt;       If TRUE, the borders of the slices will be also drawn.</span></span>
<span><span class="co">#&gt;       - title: The global (column) title of the heatmap</span></span>
<span><span class="co">#&gt;       - column_title: A character string/vector of the column name(s) to use as the title of the column group annotation.</span></span>
<span><span class="co">#&gt;       - row_title: A character string/vector of the column name(s) to use as the title of the row group annotation.</span></span>
<span><span class="co">#&gt;       - na_col: A character string specifying the color for missing values.</span></span>
<span><span class="co">#&gt;       The default is "grey85".</span></span>
<span><span class="co">#&gt;       - row_names_side: A character string specifying the side of the row names.</span></span>
<span><span class="co">#&gt;       The default is "right".</span></span>
<span><span class="co">#&gt;       - column_names_side: A character string specifying the side of the column names.</span></span>
<span><span class="co">#&gt;       The default is "bottom".</span></span>
<span><span class="co">#&gt;       - column_annotation: A character string/vector of the column name(s) to use as the column annotation.</span></span>
<span><span class="co">#&gt;       Or a list with the keys as the names of the annotation and the values as the column names.</span></span>
<span><span class="co">#&gt;       - column_annotation_side: A character string specifying the side of the column annotation.</span></span>
<span><span class="co">#&gt;       Could be a list with the keys as the names of the annotation and the values as the sides.</span></span>
<span><span class="co">#&gt;       - column_annotation_palette: A character string specifying the palette of the column annotation.</span></span>
<span><span class="co">#&gt;       The default is "Paired".</span></span>
<span><span class="co">#&gt;       Could be a list with the keys as the names of the annotation and the values as the palettes.</span></span>
<span><span class="co">#&gt;       - column_annotation_palcolor: A character vector of colors to override the palette of the column annotation.</span></span>
<span><span class="co">#&gt;       Could be a list with the keys as the names of the annotation and the values as the palcolors.</span></span>
<span><span class="co">#&gt;       - column_annotation_type: A character string specifying the type of the column annotation.</span></span>
<span><span class="co">#&gt;       The default is "auto". Other options are "simple", "pie", "ring", "bar", "violin", "boxplot", "density".</span></span>
<span><span class="co">#&gt;       Could be a list with the keys as the names of the annotation and the values as the types.</span></span>
<span><span class="co">#&gt;       If the type is "auto", the type will be determined by the type and number of the column data.</span></span>
<span><span class="co">#&gt;       - column_annotation_params: A list of parameters passed to the annotation function.</span></span>
<span><span class="co">#&gt;       Could be a list with the keys as the names of the annotation and the values as the parameters passed to the annotation function. For the parameters for names (columns_by, rows_by, columns_split_by, rows_split_by), the key should be "name.(name)", where (name) is the name of the annotation.</span></span>
<span><span class="co">#&gt;       See anno_pie(), anno_ring(), anno_bar(), anno_violin(), anno_boxplot(), anno_density(), anno_simple(), anno_points() and anno_lines() for the parameters of each annotation function.</span></span>
<span><span class="co">#&gt;       - column_annotation_agg: A function to aggregate the values in the column annotation.</span></span>
<span><span class="co">#&gt;       - row_annotation: A character string/vector of the column name(s) to use as the row annotation.</span></span>
<span><span class="co">#&gt;       Or a list with the keys as the names of the annotation and the values as the column names.</span></span>
<span><span class="co">#&gt;       - row_annotation_side: A character string specifying the side of the row annotation.</span></span>
<span><span class="co">#&gt;       Could be a list with the keys as the names of the annotation and the values as the sides.</span></span>
<span><span class="co">#&gt;       - row_annotation_palette: A character string specifying the palette of the row annotation.</span></span>
<span><span class="co">#&gt;       The default is "Paired".</span></span>
<span><span class="co">#&gt;       Could be a list with the keys as the names of the annotation and the values as the palettes.</span></span>
<span><span class="co">#&gt;       - row_annotation_palcolor: A character vector of colors to override the palette of the row annotation.</span></span>
<span><span class="co">#&gt;       Could be a list with the keys as the names of the annotation and the values as the palcolors.</span></span>
<span><span class="co">#&gt;       - row_annotation_type: A character string specifying the type of the row annotation.</span></span>
<span><span class="co">#&gt;       The default is "auto". Other options are "simple", "pie", "ring", "bar", "violin", "boxplot", "density".</span></span>
<span><span class="co">#&gt;       Could be a list with the keys as the names of the annotation and the values as the types.</span></span>
<span><span class="co">#&gt;       If the type is "auto", the type will be determined by the type and number of the row data.</span></span>
<span><span class="co">#&gt;       - row_annotation_params: A list of parameters passed to the annotation function.</span></span>
<span><span class="co">#&gt;       Could be a list with the keys as the names of the annotation and the values as the parameters.</span></span>
<span><span class="co">#&gt;       Same as column_annotation_params.</span></span>
<span><span class="co">#&gt;       - row_annotation_agg: A function to aggregate the values in the row annotation.</span></span>
<span><span class="co">#&gt;       - flip: A logical value indicating whether to flip the heatmap.</span></span>
<span><span class="co">#&gt;       The idea is that, you can simply set flip = TRUE to flip the heatmap.</span></span>
<span><span class="co">#&gt;       You don't need to swap the arguments related to rows and columns, except those you specify via ...</span></span>
<span><span class="co">#&gt;       that are passed to ComplexHeatmap::Heatmap() directly.</span></span>
<span><span class="co">#&gt;       - alpha: A numeric value between 0 and 1 specifying the transparency of the heatmap cells.</span></span>
<span><span class="co">#&gt;       - seed: The random seed to use. Default is 8525.</span></span>
<span><span class="co">#&gt;       - layer_fun_callback: A function to add additional layers to the heatmap.</span></span>
<span><span class="co">#&gt;       The function should have the following arguments: j, i, x, y, w, h, fill, sr and sc.</span></span>
<span><span class="co">#&gt;       Please also refer to the layer_fun argument in ComplexHeatmap::Heatmap.</span></span>
<span><span class="co">#&gt;       - cell_type: A character string specifying the type of the heatmap cells.</span></span>
<span><span class="co">#&gt;       The default is values. Other options are "bars", "label", "dot", "violin", "boxplot".</span></span>
<span><span class="co">#&gt;       Note that for pie chart, the values under columns specified by rows will not be used directly. Instead, the values</span></span>
<span><span class="co">#&gt;       will just be counted in different pie_group_by groups. NA values will not be counted.</span></span>
<span><span class="co">#&gt;       - cell_agg: A function to aggregate the values in the cell, for the cell type "tile" and "label".</span></span>
<span><span class="co">#&gt;       The default is mean.</span></span>
<span><span class="co">#&gt;       - combine: Whether to combine the plots into one when facet is FALSE. Default is TRUE.</span></span>
<span><span class="co">#&gt;       - nrow: A numeric value specifying the number of rows in the facet.</span></span>
<span><span class="co">#&gt;       - ncol: A numeric value specifying the number of columns in the facet.</span></span>
<span><span class="co">#&gt;       - byrow: A logical value indicating whether to fill the plots by row.</span></span>
<span><span class="co">#&gt;       - axes: A string specifying how axes should be treated. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'keep' will retain all axes in individual plots.</span></span>
<span><span class="co">#&gt;        'collect' will remove duplicated axes when placed in the same run of rows or columns of the layout.</span></span>
<span><span class="co">#&gt;        'collect_x' and 'collect_y' will remove duplicated x-axes in the columns or duplicated y-axes in the rows respectively.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - axis_titles: A string specifying how axis titltes should be treated. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'keep' will retain all axis titles in individual plots.</span></span>
<span><span class="co">#&gt;        'collect' will remove duplicated titles in one direction and merge titles in the opposite direction.</span></span>
<span><span class="co">#&gt;        'collect_x' and 'collect_y' control this for x-axis titles and y-axis titles respectively.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - guides: A string specifying how guides should be treated in the layout. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'collect' will collect guides below to the given nesting level, removing duplicates.</span></span>
<span><span class="co">#&gt;        'keep' will stop collection at this level and let guides be placed alongside their plot.</span></span>
<span><span class="co">#&gt;        'auto' will allow guides to be collected if a upper level tries, but place them alongside the plot if not.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - design: Specification of the location of areas in the layout, passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE. When specified, nrow, ncol, and byrow are ignored.</span></span>
<span><span class="co">#&gt;       See patchwork::wrap_plots() for more details.</span></span>
<span><span class="co">#&gt;   For SankeyPlot, see plotthis::SankeyPlot().</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;       [...] can be:</span></span>
<span><span class="co">#&gt;       - in_form: A character string to specify the format of the data.</span></span>
<span><span class="co">#&gt;       Possible values are "auto", "long", "lodes", "wide", "alluvia", and "counts".</span></span>
<span><span class="co">#&gt;       - x: A character string specifying the column name of the data frame to plot for the x-axis.</span></span>
<span><span class="co">#&gt;       - x_sep: A character string to concatenate the columns in x, if multiple columns are provided.</span></span>
<span><span class="co">#&gt;       - y: A character string specifying the column name of the data frame to plot for the y-axis.</span></span>
<span><span class="co">#&gt;       - stratum: A character string of the column name to group the nodes for each x.</span></span>
<span><span class="co">#&gt;       See data for more details.</span></span>
<span><span class="co">#&gt;       - stratum_sep: A character string to concatenate the columns in stratum, if multiple columns are provided.</span></span>
<span><span class="co">#&gt;       - alluvium: A character string of the column name to define the links.</span></span>
<span><span class="co">#&gt;       See data for more details.</span></span>
<span><span class="co">#&gt;       - alluvium_sep: A character string to concatenate the columns in alluvium, if multiple columns are provided.</span></span>
<span><span class="co">#&gt;       - split_by_sep: The separator for multiple split_by columns. See split_by</span></span>
<span><span class="co">#&gt;       - keep_empty: One of FALSE, TRUE and "level". It can also take a named list to specify</span></span>
<span><span class="co">#&gt;       different behavior for different columns. Without a named list, the behavior applies to the</span></span>
<span><span class="co">#&gt;       categorical/character columns used on the plot, for example, the x, group_by, fill_by, etc.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       FALSE (default): Drop empty factor levels from the data before plotting.</span></span>
<span><span class="co">#&gt;       TRUE: Keep empty factor levels and show them as a separate category in the plot.</span></span>
<span><span class="co">#&gt;       "level": Keep empty factor levels, but do not show them in the plot.</span></span>
<span><span class="co">#&gt;       But they will be assigned colors from the palette to maintain consistency across multiple plots.</span></span>
<span><span class="co">#&gt;       Alias: levels</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - flow: A logical value to use ggalluvial::geom_flow instead of ggalluvial::geom_alluvium.</span></span>
<span><span class="co">#&gt;       - expand: The values to expand the x and y axes. It is like CSS padding.</span></span>
<span><span class="co">#&gt;       When a single value is provided, it is used for both axes on both sides.</span></span>
<span><span class="co">#&gt;       When two values are provided, the first value is used for the top/bottom side and the second value is used for the left/right side.</span></span>
<span><span class="co">#&gt;       When three values are provided, the first value is used for the top side, the second value is used for the left/right side, and the third value is used for the bottom side.</span></span>
<span><span class="co">#&gt;       When four values are provided, the values are used for the top, right, bottom, and left sides, respectively.</span></span>
<span><span class="co">#&gt;       You can also use a named vector to specify the values for each side.</span></span>
<span><span class="co">#&gt;       When the axis is discrete, the values will be applied as 'add' to the 'expansion' function.</span></span>
<span><span class="co">#&gt;       When the axis is continuous, the values will be applied as 'mult' to the 'expansion' function.</span></span>
<span><span class="co">#&gt;       See also https://ggplot2.tidyverse.org/reference/expansion.html</span></span>
<span><span class="co">#&gt;       - nodes_legend: Controls how the legend of nodes will be shown. Possible values are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        "merge": Merge the legends of nodes. That is only one legend will be shown for all nodes.</span></span>
<span><span class="co">#&gt;        "separate": Show the legends of nodes separately. That is, nodes on each x will have their own legend.</span></span>
<span><span class="co">#&gt;        "none": Do not show the legend of nodes.</span></span>
<span><span class="co">#&gt;        "auto": Automatically determine how to show the legend.</span></span>
<span><span class="co">#&gt;       When nodes_label is TRUE, "none" will apply.</span></span>
<span><span class="co">#&gt;       When nodes_label is FALSE, and if stratum is the same as links_fill_by, "none" will apply.</span></span>
<span><span class="co">#&gt;       If there is any overlapping values between the nodes on different x,</span></span>
<span><span class="co">#&gt;       "merge" will apply. Otherwise, "separate" will apply.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - nodes_color: A character string to color the nodes.</span></span>
<span><span class="co">#&gt;       Use a special value ".fill" to use the same color as the fill.</span></span>
<span><span class="co">#&gt;       - links_fill_by: A character string of the column name to fill the links.</span></span>
<span><span class="co">#&gt;       - links_fill_by_sep: A character string to concatenate the columns in links_fill_by, if multiple columns are provided.</span></span>
<span><span class="co">#&gt;       - links_name: A character string to name the legend of links.</span></span>
<span><span class="co">#&gt;       - links_color: A character string to color the borders of links.</span></span>
<span><span class="co">#&gt;       Use a special value ".fill" to use the same color as the fill.</span></span>
<span><span class="co">#&gt;       - nodes_palette: A character string to specify the palette of nodes fill.</span></span>
<span><span class="co">#&gt;       - nodes_palcolor: A character vector to specify the colors of nodes fill.</span></span>
<span><span class="co">#&gt;       - nodes_alpha: A numeric value to specify the transparency of nodes fill.</span></span>
<span><span class="co">#&gt;       - nodes_label: A logical value to show the labels on the nodes.</span></span>
<span><span class="co">#&gt;       - nodes_label_miny: A numeric value to specify the minimum y (frequency) to show the labels.</span></span>
<span><span class="co">#&gt;       - nodes_width: A numeric value to specify the width of nodes.</span></span>
<span><span class="co">#&gt;       - links_palette: A character string to specify the palette of links fill.</span></span>
<span><span class="co">#&gt;       - links_palcolor: A character vector to specify the colors of links fill.</span></span>
<span><span class="co">#&gt;       - links_alpha: A numeric value to specify the transparency of links fill.</span></span>
<span><span class="co">#&gt;       - legend.box: A character string to specify the box of the legend, either "vertical" or "horizontal".</span></span>
<span><span class="co">#&gt;       - aspect.ratio: A numeric value specifying the aspect ratio of the plot.</span></span>
<span><span class="co">#&gt;       - legend.position: A character string specifying the position of the legend.</span></span>
<span><span class="co">#&gt;       if waiver(), for single groups, the legend will be "none", otherwise "right".</span></span>
<span><span class="co">#&gt;       - legend.direction: A character string specifying the direction of the legend.</span></span>
<span><span class="co">#&gt;       - flip: A logical value to flip the plot.</span></span>
<span><span class="co">#&gt;       - theme: A character string or a theme class (i.e. ggplot2::theme_classic) specifying the theme to use.</span></span>
<span><span class="co">#&gt;       Default is "theme_this".</span></span>
<span><span class="co">#&gt;       - theme_args: A list of arguments to pass to the theme function.</span></span>
<span><span class="co">#&gt;       - title: A character string specifying the title of the plot.</span></span>
<span><span class="co">#&gt;       A function can be used to generate the title based on the default title.</span></span>
<span><span class="co">#&gt;       This is useful when split_by is used and the title needs to be dynamic.</span></span>
<span><span class="co">#&gt;       - subtitle: A character string specifying the subtitle of the plot.</span></span>
<span><span class="co">#&gt;       - xlab: A character string specifying the x-axis label.</span></span>
<span><span class="co">#&gt;       - ylab: A character string specifying the y-axis label.</span></span>
<span><span class="co">#&gt;       - facet_scales: Whether to scale the axes of facets. Default is "fixed"</span></span>
<span><span class="co">#&gt;       Other options are "free", "free_x", "free_y". See ggplot2::facet_wrap</span></span>
<span><span class="co">#&gt;       - facet_ncol: A numeric value specifying the number of columns in the facet.</span></span>
<span><span class="co">#&gt;       When facet_by is a single column and facet_wrap is used.</span></span>
<span><span class="co">#&gt;       - facet_nrow: A numeric value specifying the number of rows in the facet.</span></span>
<span><span class="co">#&gt;       When facet_by is a single column and facet_wrap is used.</span></span>
<span><span class="co">#&gt;       - facet_byrow: A logical value indicating whether to fill the plots by row. Default is TRUE.</span></span>
<span><span class="co">#&gt;       - seed: The random seed to use. Default is 8525.</span></span>
<span><span class="co">#&gt;       - combine: Whether to combine the plots into one when facet is FALSE. Default is TRUE.</span></span>
<span><span class="co">#&gt;       - nrow: A numeric value specifying the number of rows in the facet.</span></span>
<span><span class="co">#&gt;       - ncol: A numeric value specifying the number of columns in the facet.</span></span>
<span><span class="co">#&gt;       - byrow: A logical value indicating whether to fill the plots by row.</span></span>
<span><span class="co">#&gt;       - axes: A string specifying how axes should be treated. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'keep' will retain all axes in individual plots.</span></span>
<span><span class="co">#&gt;        'collect' will remove duplicated axes when placed in the same run of rows or columns of the layout.</span></span>
<span><span class="co">#&gt;        'collect_x' and 'collect_y' will remove duplicated x-axes in the columns or duplicated y-axes in the rows respectively.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - axis_titles: A string specifying how axis titltes should be treated. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'keep' will retain all axis titles in individual plots.</span></span>
<span><span class="co">#&gt;        'collect' will remove duplicated titles in one direction and merge titles in the opposite direction.</span></span>
<span><span class="co">#&gt;        'collect_x' and 'collect_y' control this for x-axis titles and y-axis titles respectively.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - guides: A string specifying how guides should be treated in the layout. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'collect' will collect guides below to the given nesting level, removing duplicates.</span></span>
<span><span class="co">#&gt;        'keep' will stop collection at this level and let guides be placed alongside their plot.</span></span>
<span><span class="co">#&gt;        'auto' will allow guides to be collected if a upper level tries, but place them alongside the plot if not.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - design: Specification of the location of areas in the layout, passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE. When specified, nrow, ncol, and byrow are ignored.</span></span>
<span><span class="co">#&gt;       See patchwork::wrap_plots() for more details.</span></span>
<span><span class="co">#&gt;   For DotPlot, see plotthis::DotPlot().</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;       [...] can be:</span></span>
<span><span class="co">#&gt;       - x: A character vector specifying the column to use for the x-axis.</span></span>
<span><span class="co">#&gt;       A numeric column is expected.</span></span>
<span><span class="co">#&gt;       - y: A character vector specifying the column to use for the y-axis.</span></span>
<span><span class="co">#&gt;       A factor/character column is expected.</span></span>
<span><span class="co">#&gt;       - x_sep: A character vector to concatenate multiple columns in x. Default is "_".</span></span>
<span><span class="co">#&gt;       - y_sep: A character vector to concatenate multiple columns in y. Default is "_".</span></span>
<span><span class="co">#&gt;       - flip: A logical value indicating whether to flip the x and y axes. Default is FALSE.</span></span>
<span><span class="co">#&gt;       - split_by_sep: The separator for multiple split_by columns. See split_by</span></span>
<span><span class="co">#&gt;       - size_name: A character vector specifying the name for the size legend.</span></span>
<span><span class="co">#&gt;       - fill_name: A character vector specifying the name for the fill legend.</span></span>
<span><span class="co">#&gt;       - fill_cutoff_name: A character vector specifying the name for the fill cutoff legend.</span></span>
<span><span class="co">#&gt;       - add_bg: A logical value indicating whether to add a background color to the plot. Default is FALSE.</span></span>
<span><span class="co">#&gt;       - bg_palette: A character vector specifying the palette for the background color. Default is "stripe".</span></span>
<span><span class="co">#&gt;       - bg_palcolor: A character vector specifying the color for the background color.</span></span>
<span><span class="co">#&gt;       - bg_alpha: A numeric value specifying the alpha for the background color. Default is 0.2.</span></span>
<span><span class="co">#&gt;       - bg_direction: A character vector specifying the direction for the background color. Default is "vertical".</span></span>
<span><span class="co">#&gt;       Other options are "horizontal". "h" and "v" are also accepted.</span></span>
<span><span class="co">#&gt;       - size_by: Which column to use as the size of the dots. It must be a numeric column.</span></span>
<span><span class="co">#&gt;       If not provided, the size will be the count of the instances for each 'y' in 'x'.</span></span>
<span><span class="co">#&gt;       For 'ScatterPlot', it can be a single numeric value to specify the size of the dots.</span></span>
<span><span class="co">#&gt;       - fill_by: Which column to use as the fill the dots. It must be a numeric column.</span></span>
<span><span class="co">#&gt;       If not provided, all dots will be filled with the same color at the middle of the palette.</span></span>
<span><span class="co">#&gt;       - fill_cutoff: A numeric value specifying the cutoff for the fill column.</span></span>
<span><span class="co">#&gt;       - fill_reverse: A logical value indicating whether to reverse the fill direction. Default is FALSE.</span></span>
<span><span class="co">#&gt;       By default, the fill direction is "up". If TRUE, the fill direction is "down".</span></span>
<span><span class="co">#&gt;       When the direction is "up", the values less than the cutoff will be filled with grey.</span></span>
<span><span class="co">#&gt;       When the direction is "down", the values greater than the cutoff will be filled with grey.</span></span>
<span><span class="co">#&gt;       - theme: A character string or a theme class (i.e. ggplot2::theme_classic) specifying the theme to use.</span></span>
<span><span class="co">#&gt;       Default is "theme_this".</span></span>
<span><span class="co">#&gt;       - theme_args: A list of arguments to pass to the theme function.</span></span>
<span><span class="co">#&gt;       - palette: A character string specifying the palette to use.</span></span>
<span><span class="co">#&gt;       A named list or vector can be used to specify the palettes for different split_by values.</span></span>
<span><span class="co">#&gt;       - palcolor: A character string specifying the color to use in the palette.</span></span>
<span><span class="co">#&gt;       A named list can be used to specify the colors for different split_by values.</span></span>
<span><span class="co">#&gt;       If some values are missing, the values from the palette will be used (palcolor will be NULL for those values).</span></span>
<span><span class="co">#&gt;       - alpha: A numeric value specifying the transparency of the plot.</span></span>
<span><span class="co">#&gt;       - facet_scales: Whether to scale the axes of facets. Default is "fixed"</span></span>
<span><span class="co">#&gt;       Other options are "free", "free_x", "free_y". See ggplot2::facet_wrap</span></span>
<span><span class="co">#&gt;       - facet_ncol: A numeric value specifying the number of columns in the facet.</span></span>
<span><span class="co">#&gt;       When facet_by is a single column and facet_wrap is used.</span></span>
<span><span class="co">#&gt;       - facet_nrow: A numeric value specifying the number of rows in the facet.</span></span>
<span><span class="co">#&gt;       When facet_by is a single column and facet_wrap is used.</span></span>
<span><span class="co">#&gt;       - facet_byrow: A logical value indicating whether to fill the plots by row. Default is TRUE.</span></span>
<span><span class="co">#&gt;       - seed: The random seed to use. Default is 8525.</span></span>
<span><span class="co">#&gt;       - aspect.ratio: A numeric value specifying the aspect ratio of the plot.</span></span>
<span><span class="co">#&gt;       - legend.position: A character string specifying the position of the legend.</span></span>
<span><span class="co">#&gt;       if waiver(), for single groups, the legend will be "none", otherwise "right".</span></span>
<span><span class="co">#&gt;       - legend.direction: A character string specifying the direction of the legend.</span></span>
<span><span class="co">#&gt;       - title: A character string specifying the title of the plot.</span></span>
<span><span class="co">#&gt;       A function can be used to generate the title based on the default title.</span></span>
<span><span class="co">#&gt;       This is useful when split_by is used and the title needs to be dynamic.</span></span>
<span><span class="co">#&gt;       - subtitle: A character string specifying the subtitle of the plot.</span></span>
<span><span class="co">#&gt;       - xlab: A character string specifying the x-axis label.</span></span>
<span><span class="co">#&gt;       - ylab: A character string specifying the y-axis label.</span></span>
<span><span class="co">#&gt;       - keep_na: A logical value or a character to replace the NA values in the data.</span></span>
<span><span class="co">#&gt;       It can also take a named list to specify different behavior for different columns.</span></span>
<span><span class="co">#&gt;       If TRUE or NA, NA values will be replaced with NA.</span></span>
<span><span class="co">#&gt;       If FALSE, NA values will be removed from the data before plotting.</span></span>
<span><span class="co">#&gt;       If a character string is provided, NA values will be replaced with the provided string.</span></span>
<span><span class="co">#&gt;       If a named vector/list is provided, the names should be the column names to apply the behavior to,</span></span>
<span><span class="co">#&gt;       and the values should be one of TRUE, FALSE, or a character string.</span></span>
<span><span class="co">#&gt;       Without a named vector/list, the behavior applies to categorical/character columns used on the plot,</span></span>
<span><span class="co">#&gt;       for example, the x, group_by, fill_by, etc.</span></span>
<span><span class="co">#&gt;       - keep_empty: One of FALSE, TRUE and "level". It can also take a named list to specify</span></span>
<span><span class="co">#&gt;       different behavior for different columns. Without a named list, the behavior applies to the</span></span>
<span><span class="co">#&gt;       categorical/character columns used on the plot, for example, the x, group_by, fill_by, etc.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       FALSE (default): Drop empty factor levels from the data before plotting.</span></span>
<span><span class="co">#&gt;       TRUE: Keep empty factor levels and show them as a separate category in the plot.</span></span>
<span><span class="co">#&gt;       "level": Keep empty factor levels, but do not show them in the plot.</span></span>
<span><span class="co">#&gt;       But they will be assigned colors from the palette to maintain consistency across multiple plots.</span></span>
<span><span class="co">#&gt;       Alias: levels</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - combine: Whether to combine the plots into one when facet is FALSE. Default is TRUE.</span></span>
<span><span class="co">#&gt;       - nrow: A numeric value specifying the number of rows in the facet.</span></span>
<span><span class="co">#&gt;       - ncol: A numeric value specifying the number of columns in the facet.</span></span>
<span><span class="co">#&gt;       - byrow: A logical value indicating whether to fill the plots by row.</span></span>
<span><span class="co">#&gt;       - axes: A string specifying how axes should be treated. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'keep' will retain all axes in individual plots.</span></span>
<span><span class="co">#&gt;        'collect' will remove duplicated axes when placed in the same run of rows or columns of the layout.</span></span>
<span><span class="co">#&gt;        'collect_x' and 'collect_y' will remove duplicated x-axes in the columns or duplicated y-axes in the rows respectively.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - axis_titles: A string specifying how axis titltes should be treated. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'keep' will retain all axis titles in individual plots.</span></span>
<span><span class="co">#&gt;        'collect' will remove duplicated titles in one direction and merge titles in the opposite direction.</span></span>
<span><span class="co">#&gt;        'collect_x' and 'collect_y' control this for x-axis titles and y-axis titles respectively.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - guides: A string specifying how guides should be treated in the layout. Passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE.</span></span>
<span><span class="co">#&gt;       Options are:</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        'collect' will collect guides below to the given nesting level, removing duplicates.</span></span>
<span><span class="co">#&gt;        'keep' will stop collection at this level and let guides be placed alongside their plot.</span></span>
<span><span class="co">#&gt;        'auto' will allow guides to be collected if a upper level tries, but place them alongside the plot if not.</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;       - design: Specification of the location of areas in the layout, passed to patchwork::wrap_plots().</span></span>
<span><span class="co">#&gt;       Only relevant when split_by is used and combine is TRUE. When specified, nrow, ncol, and byrow are ignored.</span></span>
<span><span class="co">#&gt;       See patchwork::wrap_plots() for more details.</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt; - value</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;   A ggplot object or a list if combine is FALSE</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt; - examples</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;   set.seed(8525)</span></span>
<span><span class="co">#&gt;   data(cellphonedb_res)</span></span>
<span><span class="co">#&gt;   CCCPlot(data = cellphonedb_res, plot_type = "network", legend.position = "none",</span></span>
<span><span class="co">#&gt;     theme = "theme_blank", theme_args = list(add_coord = FALSE))</span></span>
<span><span class="co">#&gt;   CCCPlot(cellphonedb_res, plot_type = "chord")</span></span>
<span><span class="co">#&gt;   CCCPlot(cellphonedb_res, plot_type = "heatmap")</span></span>
<span><span class="co">#&gt;   CCCPlot(cellphonedb_res, plot_type = "dot",</span></span>
<span><span class="co">#&gt;     magnitude_agg = mean, magnitude_name = "Average Interaction Strength")</span></span>
<span><span class="co">#&gt;   CCCPlot(cellphonedb_res, plot_type = "sankey")</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;   cellphonedb_res_sub &lt;- cellphonedb_res[</span></span>
<span><span class="co">#&gt;     cellphonedb_res$source %in% c("Dendritic", "CD14+ Monocyte"),]</span></span>
<span><span class="co">#&gt;   CCCPlot(cellphonedb_res_sub, plot_type = "dot", method = "interaction")</span></span>
<span><span class="co">#&gt;   CCCPlot(cellphonedb_res_sub, plot_type = "network", method = "interaction",</span></span>
<span><span class="co">#&gt;     node_size_by = 1)</span></span>
<span><span class="co">#&gt;   CCCPlot(cellphonedb_res_sub, plot_type = "heatmap", method = "interaction",</span></span>
<span><span class="co">#&gt;     palette = "Reds")</span></span>
<span><span class="co">#&gt;   CCCPlot(cellphonedb_res_sub, plot_type = "box", method = "interaction")</span></span>
<span><span class="co">#&gt;   CCCPlot(cellphonedb_res_sub, plot_type = "violin", method = "interaction",</span></span>
<span><span class="co">#&gt;     add_box = TRUE)</span></span>
<span><span class="co">#&gt;   CCCPlot(cellphonedb_res_sub, plot_type = "ridge", method = "interaction")</span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt;   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; You must code in the programming language 'R' to answer this prompt.</span></span>
<span><span class="co">#&gt; You can use functions from these packages: scplotter.</span></span>
<span><span class="co">#&gt; You may not install or load any additional packages.</span></span>
<span><span class="co">#&gt; These objects already exist in the R session:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Object_name, Type</span></span>
<span><span class="co">#&gt; cellphonedb_res, data.frame.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Do not define these objects in your R code.</span></span></code></pre>
<pre><code><span><span class="co">#&gt; <span style="font-weight: bold;">--- Receiving response from LLM provider: ---</span></span></span></code></pre>
<pre><code><span><span class="co">#&gt; ```r</span></span>
<span><span class="co">#&gt; CCCPlot(cellphonedb_res, plot_type = "network", legend.position = "none")</span></span>
<span><span class="co">#&gt; ```</span></span></code></pre>
<pre><code><span><span class="co">#&gt; Code ran:</span></span>
<span><span class="co">#&gt; CCCPlot(cellphonedb_res, plot_type = "network", legend.position = "none")</span></span></code></pre>
<p><img src="Visualizing_data_with_LLMs_files/figure-html/unnamed-chunk-11-1.png" class="r-plt" alt="" width="672"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Panwen Wang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
